{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7fc43f2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in ./.venv/lib/python3.13/site-packages (2.3.3)\n",
      "Requirement already satisfied: openpyxl in ./.venv/lib/python3.13/site-packages (3.1.5)\n",
      "Requirement already satisfied: metapub in ./.venv/lib/python3.13/site-packages (0.6.4)\n",
      "Requirement already satisfied: transformers in ./.venv/lib/python3.13/site-packages (4.57.1)\n",
      "Requirement already satisfied: torch in ./.venv/lib/python3.13/site-packages (2.9.0)\n",
      "Requirement already satisfied: torchvision in ./.venv/lib/python3.13/site-packages (0.24.0)\n",
      "Requirement already satisfied: tensorflow in ./.venv/lib/python3.13/site-packages (2.20.0)\n",
      "Requirement already satisfied: ragas==0.3.7 in ./.venv/lib/python3.13/site-packages (0.3.7)\n",
      "Requirement already satisfied: langchain-openai in ./.venv/lib/python3.13/site-packages (1.0.1)\n",
      "Requirement already satisfied: langchain in ./.venv/lib/python3.13/site-packages (1.0.2)\n",
      "Requirement already satisfied: openai in ./.venv/lib/python3.13/site-packages (1.109.1)\n",
      "Requirement already satisfied: tiktoken in ./.venv/lib/python3.13/site-packages (0.12.0)\n",
      "Requirement already satisfied: python-dotenv in ./.venv/lib/python3.13/site-packages (1.2.1)\n",
      "Requirement already satisfied: tf-keras in ./.venv/lib/python3.13/site-packages (2.20.1)\n",
      "Requirement already satisfied: scikit-learn in ./.venv/lib/python3.13/site-packages (1.7.2)\n",
      "Requirement already satisfied: biopython in ./.venv/lib/python3.13/site-packages (1.86)\n",
      "Requirement already satisfied: beautifulsoup4 in ./.venv/lib/python3.13/site-packages (4.14.2)\n",
      "Requirement already satisfied: lxml in ./.venv/lib/python3.13/site-packages (6.0.2)\n",
      "Requirement already satisfied: textstat in ./.venv/lib/python3.13/site-packages (0.7.10)\n",
      "Requirement already satisfied: nest_asyncio in ./.venv/lib/python3.13/site-packages (1.6.0)\n",
      "Requirement already satisfied: numpy<3.0.0,>=1.21.0 in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (2.3.4)\n",
      "Requirement already satisfied: datasets>=4.0.0 in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (4.3.0)\n",
      "Requirement already satisfied: pydantic>=2.0.0 in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (2.12.3)\n",
      "Requirement already satisfied: appdirs in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (1.4.4)\n",
      "Requirement already satisfied: diskcache>=5.6.3 in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (5.6.3)\n",
      "Requirement already satisfied: typer in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (0.20.0)\n",
      "Requirement already satisfied: rich in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (14.2.0)\n",
      "Requirement already satisfied: tqdm in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (4.67.1)\n",
      "Requirement already satisfied: instructor in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (1.12.0)\n",
      "Requirement already satisfied: gitpython in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (3.1.45)\n",
      "Requirement already satisfied: pillow>=10.4.0 in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (12.0.0)\n",
      "Requirement already satisfied: networkx in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (3.5)\n",
      "Requirement already satisfied: scikit-network in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (0.33.3)\n",
      "Requirement already satisfied: langchain-core in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (1.0.1)\n",
      "Requirement already satisfied: langchain-community in ./.venv/lib/python3.13/site-packages (from ragas==0.3.7) (0.4.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in ./.venv/lib/python3.13/site-packages (from pandas) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in ./.venv/lib/python3.13/site-packages (from pandas) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in ./.venv/lib/python3.13/site-packages (from pandas) (2025.2)\n",
      "Requirement already satisfied: et-xmlfile in ./.venv/lib/python3.13/site-packages (from openpyxl) (2.0.0)\n",
      "Requirement already satisfied: setuptools in ./.venv/lib/python3.13/site-packages (from metapub) (80.9.0)\n",
      "Requirement already satisfied: lxml_html_clean in ./.venv/lib/python3.13/site-packages (from metapub) (0.4.3)\n",
      "Requirement already satisfied: requests in ./.venv/lib/python3.13/site-packages (from metapub) (2.32.5)\n",
      "Requirement already satisfied: eutils in ./.venv/lib/python3.13/site-packages (from metapub) (0.6.0)\n",
      "Requirement already satisfied: habanero in ./.venv/lib/python3.13/site-packages (from metapub) (2.3.0)\n",
      "Requirement already satisfied: tabulate in ./.venv/lib/python3.13/site-packages (from metapub) (0.9.0)\n",
      "Requirement already satisfied: cssselect in ./.venv/lib/python3.13/site-packages (from metapub) (1.3.0)\n",
      "Requirement already satisfied: unidecode in ./.venv/lib/python3.13/site-packages (from metapub) (1.4.0)\n",
      "Requirement already satisfied: docopt in ./.venv/lib/python3.13/site-packages (from metapub) (0.6.2)\n",
      "Requirement already satisfied: six in ./.venv/lib/python3.13/site-packages (from metapub) (1.17.0)\n",
      "Requirement already satisfied: coloredlogs in ./.venv/lib/python3.13/site-packages (from metapub) (15.0.1)\n",
      "Requirement already satisfied: python-Levenshtein in ./.venv/lib/python3.13/site-packages (from metapub) (0.27.1)\n",
      "Requirement already satisfied: pyyaml in ./.venv/lib/python3.13/site-packages (from metapub) (6.0.3)\n",
      "Requirement already satisfied: filelock in ./.venv/lib/python3.13/site-packages (from transformers) (3.20.0)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.34.0 in ./.venv/lib/python3.13/site-packages (from transformers) (0.36.0)\n",
      "Requirement already satisfied: packaging>=20.0 in ./.venv/lib/python3.13/site-packages (from transformers) (25.0)\n",
      "Requirement already satisfied: regex!=2019.12.17 in ./.venv/lib/python3.13/site-packages (from transformers) (2025.10.23)\n",
      "Requirement already satisfied: tokenizers<=0.23.0,>=0.22.0 in ./.venv/lib/python3.13/site-packages (from transformers) (0.22.1)\n",
      "Requirement already satisfied: safetensors>=0.4.3 in ./.venv/lib/python3.13/site-packages (from transformers) (0.6.2)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in ./.venv/lib/python3.13/site-packages (from huggingface-hub<1.0,>=0.34.0->transformers) (2025.9.0)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in ./.venv/lib/python3.13/site-packages (from huggingface-hub<1.0,>=0.34.0->transformers) (4.15.0)\n",
      "Requirement already satisfied: hf-xet<2.0.0,>=1.1.3 in ./.venv/lib/python3.13/site-packages (from huggingface-hub<1.0,>=0.34.0->transformers) (1.2.0)\n",
      "Requirement already satisfied: sympy>=1.13.3 in ./.venv/lib/python3.13/site-packages (from torch) (1.14.0)\n",
      "Requirement already satisfied: jinja2 in ./.venv/lib/python3.13/site-packages (from torch) (3.1.6)\n",
      "Requirement already satisfied: absl-py>=1.0.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (2.3.1)\n",
      "Requirement already satisfied: astunparse>=1.6.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (1.6.3)\n",
      "Requirement already satisfied: flatbuffers>=24.3.25 in ./.venv/lib/python3.13/site-packages (from tensorflow) (25.9.23)\n",
      "Requirement already satisfied: gast!=0.5.0,!=0.5.1,!=0.5.2,>=0.2.1 in ./.venv/lib/python3.13/site-packages (from tensorflow) (0.6.0)\n",
      "Requirement already satisfied: google_pasta>=0.1.1 in ./.venv/lib/python3.13/site-packages (from tensorflow) (0.2.0)\n",
      "Requirement already satisfied: libclang>=13.0.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (18.1.1)\n",
      "Requirement already satisfied: opt_einsum>=2.3.2 in ./.venv/lib/python3.13/site-packages (from tensorflow) (3.4.0)\n",
      "Requirement already satisfied: protobuf>=5.28.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (6.33.0)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (3.2.0)\n",
      "Requirement already satisfied: wrapt>=1.11.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (2.0.0)\n",
      "Requirement already satisfied: grpcio<2.0,>=1.24.3 in ./.venv/lib/python3.13/site-packages (from tensorflow) (1.76.0)\n",
      "Requirement already satisfied: tensorboard~=2.20.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (2.20.0)\n",
      "Requirement already satisfied: keras>=3.10.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (3.12.0)\n",
      "Requirement already satisfied: h5py>=3.11.0 in ./.venv/lib/python3.13/site-packages (from tensorflow) (3.15.1)\n",
      "Requirement already satisfied: ml_dtypes<1.0.0,>=0.5.1 in ./.venv/lib/python3.13/site-packages (from tensorflow) (0.5.3)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in ./.venv/lib/python3.13/site-packages (from requests->metapub) (3.4.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in ./.venv/lib/python3.13/site-packages (from requests->metapub) (3.11)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in ./.venv/lib/python3.13/site-packages (from requests->metapub) (2.5.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in ./.venv/lib/python3.13/site-packages (from requests->metapub) (2025.10.5)\n",
      "Requirement already satisfied: markdown>=2.6.8 in ./.venv/lib/python3.13/site-packages (from tensorboard~=2.20.0->tensorflow) (3.9)\n",
      "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in ./.venv/lib/python3.13/site-packages (from tensorboard~=2.20.0->tensorflow) (0.7.2)\n",
      "Requirement already satisfied: werkzeug>=1.0.1 in ./.venv/lib/python3.13/site-packages (from tensorboard~=2.20.0->tensorflow) (3.1.3)\n",
      "Requirement already satisfied: anyio<5,>=3.5.0 in ./.venv/lib/python3.13/site-packages (from openai) (4.11.0)\n",
      "Requirement already satisfied: distro<2,>=1.7.0 in ./.venv/lib/python3.13/site-packages (from openai) (1.9.0)\n",
      "Requirement already satisfied: httpx<1,>=0.23.0 in ./.venv/lib/python3.13/site-packages (from openai) (0.28.1)\n",
      "Requirement already satisfied: jiter<1,>=0.4.0 in ./.venv/lib/python3.13/site-packages (from openai) (0.10.0)\n",
      "Requirement already satisfied: sniffio in ./.venv/lib/python3.13/site-packages (from openai) (1.3.1)\n",
      "Requirement already satisfied: httpcore==1.* in ./.venv/lib/python3.13/site-packages (from httpx<1,>=0.23.0->openai) (1.0.9)\n",
      "Requirement already satisfied: h11>=0.16 in ./.venv/lib/python3.13/site-packages (from httpcore==1.*->httpx<1,>=0.23.0->openai) (0.16.0)\n",
      "Requirement already satisfied: jsonpatch<2.0.0,>=1.33.0 in ./.venv/lib/python3.13/site-packages (from langchain-core->ragas==0.3.7) (1.33)\n",
      "Requirement already satisfied: langsmith<1.0.0,>=0.3.45 in ./.venv/lib/python3.13/site-packages (from langchain-core->ragas==0.3.7) (0.4.38)\n",
      "Requirement already satisfied: tenacity!=8.4.0,<10.0.0,>=8.1.0 in ./.venv/lib/python3.13/site-packages (from langchain-core->ragas==0.3.7) (9.1.2)\n",
      "Requirement already satisfied: jsonpointer>=1.9 in ./.venv/lib/python3.13/site-packages (from jsonpatch<2.0.0,>=1.33.0->langchain-core->ragas==0.3.7) (3.0.0)\n",
      "Requirement already satisfied: orjson>=3.9.14 in ./.venv/lib/python3.13/site-packages (from langsmith<1.0.0,>=0.3.45->langchain-core->ragas==0.3.7) (3.11.4)\n",
      "Requirement already satisfied: requests-toolbelt>=1.0.0 in ./.venv/lib/python3.13/site-packages (from langsmith<1.0.0,>=0.3.45->langchain-core->ragas==0.3.7) (1.0.0)\n",
      "Requirement already satisfied: zstandard>=0.23.0 in ./.venv/lib/python3.13/site-packages (from langsmith<1.0.0,>=0.3.45->langchain-core->ragas==0.3.7) (0.25.0)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in ./.venv/lib/python3.13/site-packages (from pydantic>=2.0.0->ragas==0.3.7) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.41.4 in ./.venv/lib/python3.13/site-packages (from pydantic>=2.0.0->ragas==0.3.7) (2.41.4)\n",
      "Requirement already satisfied: typing-inspection>=0.4.2 in ./.venv/lib/python3.13/site-packages (from pydantic>=2.0.0->ragas==0.3.7) (0.4.2)\n",
      "Requirement already satisfied: langgraph<1.1.0,>=1.0.0 in ./.venv/lib/python3.13/site-packages (from langchain) (1.0.1)\n",
      "Requirement already satisfied: langgraph-checkpoint<4.0.0,>=2.1.0 in ./.venv/lib/python3.13/site-packages (from langgraph<1.1.0,>=1.0.0->langchain) (3.0.0)\n",
      "Requirement already satisfied: langgraph-prebuilt<1.1.0,>=1.0.0 in ./.venv/lib/python3.13/site-packages (from langgraph<1.1.0,>=1.0.0->langchain) (1.0.1)\n",
      "Requirement already satisfied: langgraph-sdk<0.3.0,>=0.2.2 in ./.venv/lib/python3.13/site-packages (from langgraph<1.1.0,>=1.0.0->langchain) (0.2.9)\n",
      "Requirement already satisfied: xxhash>=3.5.0 in ./.venv/lib/python3.13/site-packages (from langgraph<1.1.0,>=1.0.0->langchain) (3.6.0)\n",
      "Requirement already satisfied: ormsgpack>=1.10.0 in ./.venv/lib/python3.13/site-packages (from langgraph-checkpoint<4.0.0,>=2.1.0->langgraph<1.1.0,>=1.0.0->langchain) (1.11.0)\n",
      "Requirement already satisfied: scipy>=1.8.0 in ./.venv/lib/python3.13/site-packages (from scikit-learn) (1.16.3)\n",
      "Requirement already satisfied: joblib>=1.2.0 in ./.venv/lib/python3.13/site-packages (from scikit-learn) (1.5.2)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in ./.venv/lib/python3.13/site-packages (from scikit-learn) (3.6.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in ./.venv/lib/python3.13/site-packages (from beautifulsoup4) (2.8)\n",
      "Requirement already satisfied: pyphen in ./.venv/lib/python3.13/site-packages (from textstat) (0.17.2)\n",
      "Requirement already satisfied: nltk in ./.venv/lib/python3.13/site-packages (from textstat) (3.9.2)\n",
      "Requirement already satisfied: wheel<1.0,>=0.23.0 in ./.venv/lib/python3.13/site-packages (from astunparse>=1.6.0->tensorflow) (0.45.1)\n",
      "Requirement already satisfied: pyarrow>=21.0.0 in ./.venv/lib/python3.13/site-packages (from datasets>=4.0.0->ragas==0.3.7) (22.0.0)\n",
      "Requirement already satisfied: dill<0.4.1,>=0.3.0 in ./.venv/lib/python3.13/site-packages (from datasets>=4.0.0->ragas==0.3.7) (0.4.0)\n",
      "Requirement already satisfied: multiprocess<0.70.17 in ./.venv/lib/python3.13/site-packages (from datasets>=4.0.0->ragas==0.3.7) (0.70.16)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in ./.venv/lib/python3.13/site-packages (from fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (3.13.2)\n",
      "Requirement already satisfied: aiohappyeyeballs>=2.5.0 in ./.venv/lib/python3.13/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (2.6.1)\n",
      "Requirement already satisfied: aiosignal>=1.4.0 in ./.venv/lib/python3.13/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (1.4.0)\n",
      "Requirement already satisfied: attrs>=17.3.0 in ./.venv/lib/python3.13/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (25.4.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in ./.venv/lib/python3.13/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (1.8.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in ./.venv/lib/python3.13/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (6.7.0)\n",
      "Requirement already satisfied: propcache>=0.2.0 in ./.venv/lib/python3.13/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (0.4.1)\n",
      "Requirement already satisfied: yarl<2.0,>=1.17.0 in ./.venv/lib/python3.13/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.9.0,>=2023.1.0->datasets>=4.0.0->ragas==0.3.7) (1.22.0)\n",
      "Requirement already satisfied: namex in ./.venv/lib/python3.13/site-packages (from keras>=3.10.0->tensorflow) (0.1.0)\n",
      "Requirement already satisfied: optree in ./.venv/lib/python3.13/site-packages (from keras>=3.10.0->tensorflow) (0.17.0)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in ./.venv/lib/python3.13/site-packages (from sympy>=1.13.3->torch) (1.3.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in ./.venv/lib/python3.13/site-packages (from werkzeug>=1.0.1->tensorboard~=2.20.0->tensorflow) (3.0.3)\n",
      "Requirement already satisfied: humanfriendly>=9.1 in ./.venv/lib/python3.13/site-packages (from coloredlogs->metapub) (10.0)\n",
      "Requirement already satisfied: gitdb<5,>=4.0.1 in ./.venv/lib/python3.13/site-packages (from gitpython->ragas==0.3.7) (4.0.12)\n",
      "Requirement already satisfied: smmap<6,>=3.0.1 in ./.venv/lib/python3.13/site-packages (from gitdb<5,>=4.0.1->gitpython->ragas==0.3.7) (5.0.2)\n",
      "Requirement already satisfied: docstring-parser<1.0,>=0.16 in ./.venv/lib/python3.13/site-packages (from instructor->ragas==0.3.7) (0.17.0)\n",
      "Requirement already satisfied: pre-commit>=4.3.0 in ./.venv/lib/python3.13/site-packages (from instructor->ragas==0.3.7) (4.3.0)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in ./.venv/lib/python3.13/site-packages (from rich->ragas==0.3.7) (4.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in ./.venv/lib/python3.13/site-packages (from rich->ragas==0.3.7) (2.19.2)\n",
      "Requirement already satisfied: click>=8.0.0 in ./.venv/lib/python3.13/site-packages (from typer->ragas==0.3.7) (8.3.0)\n",
      "Requirement already satisfied: shellingham>=1.3.0 in ./.venv/lib/python3.13/site-packages (from typer->ragas==0.3.7) (1.5.4)\n",
      "Requirement already satisfied: mdurl~=0.1 in ./.venv/lib/python3.13/site-packages (from markdown-it-py>=2.2.0->rich->ragas==0.3.7) (0.1.2)\n",
      "Requirement already satisfied: cfgv>=2.0.0 in ./.venv/lib/python3.13/site-packages (from pre-commit>=4.3.0->instructor->ragas==0.3.7) (3.4.0)\n",
      "Requirement already satisfied: identify>=1.0.0 in ./.venv/lib/python3.13/site-packages (from pre-commit>=4.3.0->instructor->ragas==0.3.7) (2.6.15)\n",
      "Requirement already satisfied: nodeenv>=0.11.1 in ./.venv/lib/python3.13/site-packages (from pre-commit>=4.3.0->instructor->ragas==0.3.7) (1.9.1)\n",
      "Requirement already satisfied: virtualenv>=20.10.0 in ./.venv/lib/python3.13/site-packages (from pre-commit>=4.3.0->instructor->ragas==0.3.7) (20.35.4)\n",
      "Requirement already satisfied: distlib<1,>=0.3.7 in ./.venv/lib/python3.13/site-packages (from virtualenv>=20.10.0->pre-commit>=4.3.0->instructor->ragas==0.3.7) (0.4.0)\n",
      "Requirement already satisfied: platformdirs<5,>=3.9.1 in ./.venv/lib/python3.13/site-packages (from virtualenv>=20.10.0->pre-commit>=4.3.0->instructor->ragas==0.3.7) (4.5.0)\n",
      "Requirement already satisfied: langchain-classic<2.0.0,>=1.0.0 in ./.venv/lib/python3.13/site-packages (from langchain-community->ragas==0.3.7) (1.0.0)\n",
      "Requirement already satisfied: SQLAlchemy<3.0.0,>=1.4.0 in ./.venv/lib/python3.13/site-packages (from langchain-community->ragas==0.3.7) (2.0.44)\n",
      "Requirement already satisfied: dataclasses-json<0.7.0,>=0.6.7 in ./.venv/lib/python3.13/site-packages (from langchain-community->ragas==0.3.7) (0.6.7)\n",
      "Requirement already satisfied: pydantic-settings<3.0.0,>=2.10.1 in ./.venv/lib/python3.13/site-packages (from langchain-community->ragas==0.3.7) (2.11.0)\n",
      "Requirement already satisfied: httpx-sse<1.0.0,>=0.4.0 in ./.venv/lib/python3.13/site-packages (from langchain-community->ragas==0.3.7) (0.4.3)\n",
      "Requirement already satisfied: marshmallow<4.0.0,>=3.18.0 in ./.venv/lib/python3.13/site-packages (from dataclasses-json<0.7.0,>=0.6.7->langchain-community->ragas==0.3.7) (3.26.1)\n",
      "Requirement already satisfied: typing-inspect<1,>=0.4.0 in ./.venv/lib/python3.13/site-packages (from dataclasses-json<0.7.0,>=0.6.7->langchain-community->ragas==0.3.7) (0.9.0)\n",
      "Requirement already satisfied: langchain-text-splitters<2.0.0,>=1.0.0 in ./.venv/lib/python3.13/site-packages (from langchain-classic<2.0.0,>=1.0.0->langchain-community->ragas==0.3.7) (1.0.0)\n",
      "Requirement already satisfied: mypy-extensions>=0.3.0 in ./.venv/lib/python3.13/site-packages (from typing-inspect<1,>=0.4.0->dataclasses-json<0.7.0,>=0.6.7->langchain-community->ragas==0.3.7) (1.1.0)\n",
      "Requirement already satisfied: Levenshtein==0.27.1 in ./.venv/lib/python3.13/site-packages (from python-Levenshtein->metapub) (0.27.1)\n",
      "Requirement already satisfied: rapidfuzz<4.0.0,>=3.9.0 in ./.venv/lib/python3.13/site-packages (from Levenshtein==0.27.1->python-Levenshtein->metapub) (3.14.1)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/t2yu/Desktop/Lay-Summary-Project/.venv/lib/python3.13/site-packages/eutils/__init__.py:4: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "2025-11-12 15:05:26 ll6684-mlaszk.ad.oicr.on.ca metapub.config[16357] WARNING NCBI_API_KEY was not set.\n",
      "/Users/t2yu/Desktop/Lay-Summary-Project/.venv/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "2025-11-12 15:06:24 ll6684-mlaszk.ad.oicr.on.ca datasets[16357] INFO TensorFlow version 2.20.0 available.\n"
     ]
    }
   ],
   "source": [
    "# Install and import necessary libraries\n",
    "\n",
    "!python3 -m pip install pandas openpyxl metapub transformers torch torchvision tensorflow ragas==0.3.7 langchain-openai langchain openai tiktoken python-dotenv tf-keras scikit-learn biopython beautifulsoup4 lxml textstat openai nest_asyncio\n",
    "import pandas as pd\n",
    "import json\n",
    "from metapub import PubMedFetcher\n",
    "from transformers import pipeline\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from functools import reduce\n",
    "import os\n",
    "from datetime import datetime\n",
    "from dotenv import load_dotenv\n",
    "from openai import OpenAI\n",
    "from Bio import Entrez\n",
    "from bs4 import BeautifulSoup\n",
    "from ragas.metrics import faithfulness\n",
    "from ragas import evaluate\n",
    "from ragas.llms import LangchainLLMWrapper\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain_core.documents import Document\n",
    "from datasets import Dataset\n",
    "import csv\n",
    "\n",
    "import time\n",
    "import textstat\n",
    "from pathlib import Path\n",
    "\n",
    "import nest_asyncio\n",
    "nest_asyncio.apply()\n",
    "\n",
    "from openai import AsyncOpenAI\n",
    "import tqdm\n",
    "import asyncio\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4cb1098b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "API Key Loaded: sk-p...\n",
      "Entrez email set: t2yu@oicr.on.ca\n",
      "Entrez API key: da05...\n",
      "Lay Summarizer Configuration: \n",
      " Number of articles: 1 \n",
      " Maximum word length: 350 \n",
      " Minimum word length: 250\n"
     ]
    }
   ],
   "source": [
    "# Load environment variables from your .env file, or export your api key\n",
    "# load_dotenv(\"\")\n",
    "openai_api_key = os.environ.get(\"OPENAI_API_KEY\")\n",
    "if not openai_api_key:\n",
    "    raise ValueError(\"API key not found.\")\n",
    "else:\n",
    "    print(f\"API Key Loaded: {openai_api_key[:4]}...\")\n",
    "\n",
    "# OpenAI client\n",
    "os.environ[\"OPENAI_API_KEY\"] = openai_api_key\n",
    "client = OpenAI(api_key=openai_api_key)\n",
    "\n",
    "# Set location of configuration file, to load configuration variables\n",
    "with open(\"./config.json\", \"r\") as config_file:\n",
    "    config = json.load(config_file)\n",
    "\n",
    "num_of_articles = config[\"num_of_articles\"]\n",
    "max_summary_length = config[\"max_summary_length\"]\n",
    "min_summary_length = config[\"min_summary_length\"]\n",
    "\n",
    "# Set an Entrez email within the config file\n",
    "fetch = PubMedFetcher()\n",
    "Entrez.email = config[\"entrez_email\"]\n",
    "Entrez.api_key = config[\"entrez_api_key\"]\n",
    "print(f\"Entrez email set: {Entrez.email}\")\n",
    "print(f\"Entrez API key: {Entrez.api_key[:4]}...\")\n",
    "print(f\"Lay Summarizer Configuration: \\n Number of articles: {num_of_articles} \\n Maximum word length: {max_summary_length} \\n Minimum word length: {min_summary_length}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d8c61905",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch full texts \n",
    "def fetch_full_texts(pmids, fetcher):\n",
    "    time.sleep(1) \n",
    "    full_texts, titles = {}, {}\n",
    "\n",
    "    print(f\"Retrieved PMIDs: {pmids}\")\n",
    "    print(f\"Number of PMIDs retrieved: {len(pmids)}\")\n",
    "\n",
    "    for pmid in pmids:\n",
    "        try:\n",
    "            article = fetcher.article_by_pmid(pmid)\n",
    "            titles[pmid] = article.title\n",
    "            \n",
    "            pmcid = article.pmc\n",
    "            if pmcid:\n",
    "                handle = Entrez.efetch(db=\"pmc\", id=pmcid, rettype=\"xml\", retmode=\"xml\")\n",
    "                xml_data = handle.read()\n",
    "                handle.close()\n",
    "\n",
    "                soup = BeautifulSoup(xml_data, features=\"xml\")\n",
    "\n",
    "                paragraphs = soup.find_all(\"p\")\n",
    "                sections = soup.find_all(\"sec\")\n",
    "\n",
    "                text_chunks = [p.get_text() for p in paragraphs]\n",
    "                if not text_chunks and sections:\n",
    "                    # Fallback to extracting from section text\n",
    "                    text_chunks = [s.get_text() for s in sections]\n",
    "\n",
    "                full_text = \"\\n\".join(p.get_text() for p in paragraphs)\n",
    "                full_texts[pmid] = full_text if full_text.strip() else \"Full text unavailable\"\n",
    "                \n",
    "            else:\n",
    "                full_texts[pmid] = \"Full text unavailable\"\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching full text for PMID {pmid}: {e}\")\n",
    "            titles[pmid] = titles.get(pmid, \"Unavailable\")\n",
    "            full_texts[pmid] = \"Full text unavailable\"\n",
    "    return titles, full_texts\n",
    "    \n",
    "\n",
    "def fetch_abstracts(pmids, fetcher):\n",
    "    abstracts = {}\n",
    "    titles = {}\n",
    "    \n",
    "    for pmid in pmids:\n",
    "        try:\n",
    "            article = fetcher.article_by_pmid(pmid)\n",
    "            titles[pmid] = article.title\n",
    "            abstracts[pmid] = article.abstract if article.abstract else \"Abstract unavailable\"\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching abstract for PMID {pmid}: {e}\")\n",
    "            titles[pmid] = titles.get(pmid, \"Unavailable\")\n",
    "            abstracts[pmid] = \"Abstract unavailable\"\n",
    "    return titles, abstracts\n",
    "\n",
    "async_client = AsyncOpenAI(api_key=openai_api_key)\n",
    "\n",
    "async def async_summarize(pmid, text, max_length):\n",
    "    if not text or text == \"Full text unavailable\":\n",
    "        return pmid, \"No text available\"\n",
    "\n",
    "    prompt = (\n",
    "        f\"Summarize this article in 250 to {max_length} words for a 2nd grade lay audience using simple words. \"\n",
    "        f\"The tone should be casual. Avoid emotive or negative language such as battle or fight. \"\n",
    "        f\"Keep words under 4 syllables and sentences under 10 words: {text}\"\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        response = await async_client.chat.completions.create(\n",
    "            model=\"gpt-4o\", #\"gpt-4-turbo\",\n",
    "            messages=[\n",
    "                {\"role\": \"system\", \"content\": \"You are a helpful assistant that summarizes medical research.\"},\n",
    "                {\"role\": \"user\", \"content\": prompt}\n",
    "            ],\n",
    "            temperature=0.5,\n",
    "            \n",
    "        )\n",
    "        content = response.choices[0].message.content.strip()\n",
    "        return pmid, content\n",
    "    except Exception as e:\n",
    "        return pmid, f\"Error summarizing: {e}\"\n",
    "\n",
    "async def gpt4_summarize_async(texts, max_length, concurrency=15):\n",
    "    semaphore = asyncio.Semaphore(concurrency)\n",
    "\n",
    "    async def limited_summary(pmid, text):\n",
    "        async with semaphore:\n",
    "            return await async_summarize(pmid, text, max_length)\n",
    "\n",
    "    tasks = [limited_summary(pmid, text) for pmid, text in texts.items()]\n",
    "    results = await asyncio.gather(*tasks)\n",
    "    return dict(results)\n",
    "\n",
    "\n",
    "def compute_readability(summary):\n",
    "    if summary and summary != \"No text available\":\n",
    "        flesch_score = textstat.flesch_reading_ease(summary)\n",
    "        grade_level = textstat.flesch_kincaid_grade(summary)\n",
    "        return flesch_score, grade_level\n",
    "    else:\n",
    "        return None, None\n",
    "\n",
    "\n",
    "def compute_ragas_faithfulness(summaries, references, prompt):\n",
    "    evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model=\"gpt-4o\"))\n",
    "    data = {\n",
    "        \"question\": [],\n",
    "        \"answer\": [],\n",
    "        \"contexts\": []\n",
    "    }\n",
    "    valid_pmids = []\n",
    "\n",
    "    for pmid, summary in summaries.items():\n",
    "        reference = references.get(pmid)\n",
    "        if (\n",
    "            summary \n",
    "            and reference \n",
    "            and summary != \"No text available\" \n",
    "            and reference != \"Full text unavailable\"\n",
    "        ):\n",
    "            data[\"question\"].append(prompt)\n",
    "            data[\"answer\"].append(summary)\n",
    "            data[\"contexts\"].append([reference])\n",
    "            valid_pmids.append(pmid)\n",
    "\n",
    "    \n",
    "    dataset = Dataset.from_dict(data)\n",
    "    results = evaluate(dataset, metrics=[faithfulness], llm=evaluator_llm)\n",
    "    return dict(zip(valid_pmids, results[\"faithfulness\"]))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f109742",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['39119834']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Use if you need to query for PMIDS\n",
    "if \"queries\" in config:\n",
    "\n",
    "    for query in config[\"queries\"]:\n",
    "            keyword = query[\"keyword\"]\n",
    "            output_file = query[\"output_file\"]\n",
    "\n",
    "            pmids = fetch.pmids_for_query(f\"{keyword} AND pubmed pmc open access[filter]\", retmax=num_of_articles)\n",
    "\n",
    "\n",
    "            print(pmids)\n",
    "\n",
    "            #writes the queried PMIDs to a txt file\n",
    "            with open(\"pmids_queried.txt\", \"w\") as f:\n",
    "                for pmid in pmids:\n",
    "                    f.write(pmid + \"\\n\")\n",
    "\n",
    "            \n",
    "else:\n",
    "    print(\"No queries found in the configuration file.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e7ad773c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 40 PMIDs from pmids_queried.txt\n"
     ]
    }
   ],
   "source": [
    "# Use if you have a file of PMIDs to summarize and calculate metrics\n",
    "if \"pmid_file\" in config:\n",
    "    pmid_file = config[\"pmid_file\"]\n",
    "    with open(pmid_file, \"r\") as f:\n",
    "        pmids = [line.strip() for line in f.readlines()]\n",
    "    print(f\"Loaded {len(pmids)} PMIDs from {pmid_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "375cbbbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{40665378: \"This article talks about kidney cancer, which is a type of cancer that starts in the kidneys. Kidneys are like little bean-shaped filters in our body that help clean our blood. Sometimes, cells in the kidneys grow too much and form a lump called a tumor. When these tumors grow and spread to other parts of the body, it's called metastatic renal cell carcinoma, or mRCC for short.\\n\\nDoctors use special medicines to help people with mRCC. These medicines are called immune checkpoint inhibitors (ICIs) and tyrosine kinase inhibitors (TKIs). They help the body's defense system work better to stop the cancer. But sometimes, these medicines stop working well because the cancer cells find ways to hide from them. One way cancer cells do this is through a process called autophagy. Autophagy is like a recycling system inside cells that helps them stay healthy. But in cancer cells, it can make them stronger and harder to treat.\\n\\nThe scientists in this study found that cancer cells in the kidneys use a lot of autophagy. They tried using a different medicine called DCMI, which can block autophagy. When they used DCMI with another medicine called Sunitinib, it worked better to stop the cancer cells from growing and spreading.\\n\\nThe scientists think that blocking autophagy might help make cancer treatments work better. They are excited about using DCMI because it might help people with kidney cancer feel better for longer. They want to keep studying this to see if it can help more people in the future.\", 40665456: 'Alright, kids! Let’s talk about a study where doctors looked at something called the \"tumor sink effect.\" This is a fancy way of saying that some medicine gets stuck in tumors, which are like little lumps in the body. This means less medicine is left for the rest of the body.\\n\\nThe study had 97 people with a type of cancer called prostate cancer. They used a special camera to see where the medicine went in their bodies. They found out that people with big tumors had more medicine stuck in those tumors. This left less medicine for healthy parts like the liver, kidneys, and blood.\\n\\nDoctors put people into three groups based on how much medicine got stuck in the tumors: small, medium, and large. They saw that when more medicine got stuck in the tumor, less was left for healthy parts.\\n\\nDoctors also looked at things like how big or small the people were and how well their kidneys worked. They found these things also changed how much medicine went to healthy parts.\\n\\nThe study showed that people with big tumors might not get enough medicine to other parts of their bodies. This is important because doctors want to make sure everyone gets the right amount of medicine.\\n\\nIn the end, the study helps doctors understand how to give the best care to people with prostate cancer. They learned that some people might need a different amount of medicine to help their bodies best. This way, everyone gets just what they need to feel better!', 40666060: 'Alright, let’s talk about a story about a young woman who got very sick. She had a condition called CAPS, which is really rare and makes lots of tiny clogs in the blood. These clogs can happen in many parts of the body, like the heart, lungs, and kidneys, and can make someone feel really unwell.\\n\\nThis young woman had some other health problems too. She had something called lupus and had clogs in her legs before. Then, she got COVID-19, which made things even trickier for her.\\n\\nDoctors did lots of tests to see what was wrong. They used special machines like CT scans to take pictures inside her body. They found clogs in her lungs and some other problems in her belly and kidneys.\\n\\nHer kidneys, which help clean the blood, were not working well. So, the doctors did more tests and found out her kidneys were very sick too.\\n\\nThe doctors gave her lots of medicine to help her feel better. They used something called plasmapheresis, which is like a special cleaning for the blood. They also gave her medicine to stop the clogs and help with the swelling.\\n\\nAfter lots of care, she started to feel better. Her face and legs stopped being puffy, and she could go home with medicine to keep her healthy.\\n\\nThis story helps doctors learn more about CAPS and how to help people who have it. It’s important because CAPS is very rare, and every story helps doctors understand it better.', 40667694: \"This study is about kidney cancer, called renal cell carcinoma (RCC). This type of cancer is tough to treat because it doesn't respond well to regular medicine. Scientists are trying to use a special treatment called immunotherapy. But sometimes this doesn't work either, and they want to find out why.\\n\\nThe scientists looked at many research papers from 2020 to 2024. They used special tools to see which authors, countries, and journals were most important in this research. They found that some authors like Guo Jianming and Moon, James J. wrote a lot about RCC. Important journals were Cancers and Frontiers in Immunology.\\n\\nThe research papers talked about how the body's immune system can help fight cancer. But sometimes the cancer tricks the immune system, so the treatment stops working. Scientists want to find new ways to make the treatment work better. They are looking at things called biomarkers. These are tiny signs in the body that can tell doctors if a treatment will work or not.\\n\\nThe scientists also found that using two or more treatments together might help. This is called combination therapy. It means using different medicines or treatments at the same time to help beat the cancer.\\n\\nSome countries like the USA, China, and France are doing a lot of research on RCC. They are trying to find out how to make treatments better and help people with this cancer live longer.\\n\\nThe study shows that finding the right biomarkers and using combination therapy are important steps. These could help doctors know which treatments will work best for each person. Scientists are working hard to understand more about RCC and how to treat it better.\", 40669903: 'Let’s talk about feeling really, really tired. Some people with a condition called chronic kidney disease, or CKD for short, feel this way a lot. CKD means the kidneys are not working as well as they should. People with CKD sometimes feel very tired even when they haven’t done much. This tiredness is called fatigue. \\n\\nDoctors and scientists wanted to learn more about how to measure this tiredness in people with CKD. They looked at 97 different studies to find out how tiredness was measured. They found that there are many tools, or ways, to ask people about their tiredness. Some tools are long, and some are short. Some ask about tiredness over the past week, and others ask about the past month. \\n\\nThe scientists found 27 different tools, but only a few were made just for people with CKD. Most tools were used only once in the studies. Two tools, called SF-36 and KDQOL-SF, were used the most. But these tools weren’t made just for people with early CKD. \\n\\nThe scientists say it’s important to have a tool just for CKD tiredness. This way, doctors can better understand how tired people feel and help them more. They also say that having a good tool will help in making plans to help people feel less tired. \\n\\nThe scientists want to make a new tool that is easy to use and helps everyone understand tiredness better. They hope this will help people with CKD feel better and enjoy their lives more.', 40671833: \"Let's talk about something called MASLD. That's a big word, but it means a kind of liver problem. When your liver has too much fat and it gets sick, that's MASLD. Doctors want to know how sick the liver is, and they look for something called fibrosis. Fibrosis is like when the liver gets a bit hard and not so squishy. \\n\\nDoctors use a special camera to take pictures of the liver. This camera uses light in a fancy way to see the liver without any paint or color. It’s called SHG/TPEF. Then, a computer looks at these pictures to tell us how much fibrosis is there. The computer gives us a number called qFibrosis. This helps doctors know how sick the liver is.\\n\\nThe study wanted to find out if the size of the liver piece they look at changes the qFibrosis number. They found out that bigger pieces (more than 20 mm long and 0.9 mm wide) give better information. It’s like when you have a bigger piece of a puzzle, you can see the picture better.\\n\\nIt doesn’t matter where the piece comes from in the liver; it’s the size that counts. So, if doctors use bigger pieces, they can tell better how sick the liver is. This helps them help people get better.\\n\\nDoctors and scientists want to make sure they always use the right size pieces so they can always tell how much fibrosis is in the liver. This way, they can help people with MASLD feel better and keep their livers healthy.\", 40671868: \"Okay, let's talk about a cool robot that helps doctors do surgery! This robot is called the Carina Platform, and it comes from a company in China named Ronovo Surgical. Doctors used this robot to help people with problems in their kidneys and prostate, which are parts inside our body that need to be healthy.\\n\\nThe doctors did 18 surgeries with this robot. They used it to help 11 people with their prostate and 7 people with their kidneys. The surgeries happened at a big hospital in Shanghai, China. The doctors wanted to see if the robot was safe and if it worked well.\\n\\nHere's how the robot helped: it has special arms that can move and help the doctors see and do things inside the body without making big cuts. This is called robotic-assisted surgery. The robot helps doctors be very careful and precise.\\n\\nThe doctors were happy because all the surgeries went well. They didn't have to switch to other ways of doing surgery, like making bigger cuts. The robot helped them finish the surgeries safely. The people who had surgery didn't have any big problems afterward.\\n\\nThe robot has different parts that can be moved around. This makes it easy for doctors to set it up for different kinds of surgeries. The robot also lets doctors see inside the body in 3D, which is like watching a movie with special glasses!\\n\\nIn the end, everyone was pleased with how the robot worked. The doctors learned a lot and think this robot can help more people in the future. They want to keep using it to make surgeries easier and safer. So, that's the story of the Carina robot helping doctors do a great job!\", 40672051: 'Hey kids! Let’s talk about something important: how doctors find out if a child has cancer. Cancer is when some cells in the body grow differently. Doctors need to find it early so they can help kids feel better.\\n\\nThis study looked at how long it takes for doctors in the UK to find out if a child has cancer. They checked how fast doctors could figure it out for almost 2,000 kids.\\n\\nHere’s what they found: \\n\\n1. **Time to Find Out**: For most kids, it took about 4 to 5 weeks from the first sign of feeling unwell to knowing it was cancer. But for some, it took much longer.\\n\\n2. **Who Sees the Doctor**: Most kids went to see a regular doctor or visited the emergency room first. Very few saw a child doctor (pediatrician) right away.\\n\\n3. **Symptoms**: Kids usually had about 4 signs, like feeling very tired or having a fever, before doctors knew it was cancer.\\n\\n4. **Age Matters**: Older kids, especially teenagers, took longer to get a diagnosis than younger kids.\\n\\n5. **Type of Cancer**: Some cancers, like bone cancer, took longer to find out than others, like kidney cancer.\\n\\n6. **No Difference by Gender or Ethnicity**: It didn’t matter if the child was a boy or girl, or what their background was; the time to find out was similar.\\n\\nThe study helps doctors and health workers know where they can do better. This way, they can help kids get the care they need faster. They also want to teach everyone more about the signs of cancer so kids can get help sooner.\\n\\nAnd guess what? There’s a new plan to make sure kids get diagnosed even quicker in the future. It’s called Child Cancer Smart, and it’s all about helping doctors and families know what to look for.\\n\\nRemember, if you feel unwell, it’s always good to tell a grown-up. They can help decide if you need to see a doctor.', 40673289: \"Doctors use special medicines called ICIs to help people with cancer. These medicines help the body's defense system, called the immune system, to find and remove cancer cells. But sometimes ICIs can make the immune system too strong, and it can bother other parts of the body. This can make people feel sick in different ways.\\n\\nOne man, who was 60 years old, had a type of cancer called kidney cancer. He was taking two ICIs named nivolumab and ipilimumab. He started feeling very tired and had trouble breathing. When he went to the hospital, doctors found that his muscles, including his heart muscle, were swollen and sore. This is called myositis and myocarditis. He also had trouble swallowing, which is a sign of something called myasthenia gravis (MG). Together, these problems are called MMM overlap syndrome.\\n\\nThe doctors gave him special medicine to help calm down his immune system. They used things like IV methylprednisolone and IV immunoglobulin. After a few days, he started feeling better, and his tests showed he was improving. He left the hospital with some medicine to take at home and will keep seeing his doctor to make sure he stays well.\\n\\nMMM overlap syndrome is rare but can be serious. Doctors need to watch for it when people take ICIs, because the signs can be tricky to spot. It's important for doctors and patients to know about these signs, so they can get help quickly if needed. As more people use ICIs, doctors will learn more about how to help people with these side effects.\", 40673609: 'Hey there, kids! Today, let\\'s learn about a special study that helps doctors understand a type of kidney cancer called KIRC. Scientists are trying to find ways to tell how people with this cancer might do in the future and how best to help them get better.\\n\\nIn our bodies, we have tiny things called mRNAs. They are like little messengers that help make proteins. Some of these mRNAs are linked to a group of proteins called TNF. These proteins are important because they can affect how cancer grows.\\n\\nThe scientists in this study looked at a bunch of mRNAs that are related to TNF. They wanted to see if these mRNAs could help them predict how people with KIRC might do. They found nine special mRNAs that seemed to be really important.\\n\\nThey used these mRNAs to make a model, like a fancy calculator, that helps predict if someone with KIRC might have a better or worse time with their cancer. They called people with a higher chance of getting sicker the \"high-risk\" group and those with a lower chance the \"low-risk\" group.\\n\\nThe scientists also looked at different ways to treat people in the high-risk group. They found that these people might need special medicines that help the body\\'s immune system fight cancer better.\\n\\nThis study is like a big puzzle. Each piece helps doctors understand more about KIRC and how to treat it. By knowing which mRNAs are important, doctors can help make sure people get the right kind of care.\\n\\nSo, this study is all about using science to help people with cancer feel better and live longer. Isn\\'t that cool?', 40673839: \"Okay, let's talk about something cool called urinary extracellular vesicles, or uEVs for short. These are tiny bubbles that float in our pee. They might help doctors learn about our health.\\n\\nScientists think uEVs could be great for finding clues about diseases. This is because they can carry special messages from our body parts. But, there are some puzzles to solve before doctors can use them in real life.\\n\\nFirst, we need to figure out how to collect and study uEVs in the same way every time. This is important so scientists can compare results and be sure they are correct. Imagine if everyone used different rules to play a game – it would be confusing!\\n\\nNext, we need better tools to see these tiny bubbles. They are super small, and we need special machines to look at them. This will help scientists understand what’s inside and what it means for our health.\\n\\nAlso, scientists and doctors need to work together. They can share ideas and make sure everyone is on the same page. This teamwork can help solve the puzzles faster.\\n\\nFinally, there are rules to follow before using uEVs in hospitals. Scientists need to show that uEVs really help in finding diseases. This means doing lots of tests and getting approval from special groups.\\n\\nIn the end, if we solve these puzzles, uEVs could become a super helpful tool for doctors. They could help find diseases early and make sure we stay healthy. Scientists are excited and working hard to make this happen!\", 40677495: \"Once upon a time, there was a young woman named Jane. Jane was 29 years old and had some trouble with her body. She felt tired and had pain on her left side. Sometimes, she even saw blood when she went to the bathroom. Jane decided to visit the doctor to find out what was going on.\\n\\nThe doctor did some special tests called a CT scan. This test took pictures inside Jane's body. The doctor saw something unusual in her left kidney and also found a bump in her bladder. Jane needed more tests to understand what was happening.\\n\\nThe doctors did a special look inside her bladder and took out the bump. They also drained some fluid from her kidney. Later, they took out her left kidney in a surgery. When they looked at the kidney and bladder bumps under a microscope, they found something called chromophobe RCC. This is a rare type of kidney bump that can sometimes spread to other parts of the body, like the bladder.\\n\\nJane's doctors worked hard to help her. They did another surgery to make sure they took out all the bad bumps. Jane felt better for a while, but she couldn't keep going to the doctor because she lived far away and didn't have much money.\\n\\nEven though Jane's bumps were taken out, they came back later. The doctors learned a lot from helping Jane. They saw that even rare things can happen, like kidney bumps spreading to the bladder. They also learned that it's important to help everyone get the care they need, no matter where they live or how much money they have.\\n\\nThis story shows how doctors and nurses work together to help people feel better. It also reminds us that taking care of our bodies and seeing the doctor when something feels wrong is very important.\", 40677502: \"Statins are medicines that help keep our hearts healthy. They do this by lowering bad stuff in our blood called cholesterol. Cholesterol can make our heart sick if there's too much. So, statins are like helpers that keep our hearts strong and happy.\\n\\nBut sometimes, statins can cause problems. These problems are rare, which means they don’t happen often. Some people might feel weak or have muscle pain. In very rare cases, statins can make muscles break down too much. This can make other parts of the body, like the liver and kidneys, not work well.\\n\\nDoctors are super careful when giving statins. They watch for signs that something might be wrong. If someone feels very weak or has dark pee, they should tell a doctor right away. Doctors can check if the statins are causing these problems.\\n\\nThere was a man who took statins for a long time. He started feeling very weak and couldn’t walk well. His doctor found out that the statins were causing his muscles to break down too much. This made his kidneys and liver not work well. The doctor stopped the statins and gave him lots of water through a drip to help his body get better.\\n\\nEven though this can happen, statins are usually safe and help many people. Doctors choose the right kind of statin for each person. They make sure it helps the heart without causing problems. If you take statins, it’s important to tell your doctor how you feel. This helps them make sure the medicine is right for you.\\n\\nStatins are good for the heart, but we have to be careful. If something feels wrong, doctors can help fix it. This way, we can keep using statins to stay strong and healthy!\", 40677703: 'Hey there! Let’s talk about a kind of cancer called kidney cancer. Most kidney cancers are called renal cell carcinoma, or RCC for short. The most common type of RCC is called clear-cell RCC. Doctors usually treat it by removing the tumor through surgery. But sometimes, the cancer can come back. So, scientists are working hard to find new ways to help people who have this cancer.\\n\\nThere are some cool new treatments called immunotherapy and targeted therapy. These treatments help the body’s own defenses, called the immune system, to find and stop cancer cells. Immunotherapy uses special medicines to make the immune system stronger. Targeted therapy uses medicines to stop cancer cells from growing by blocking signals they need to grow.\\n\\nOne type of immunotherapy uses medicines called immune checkpoint inhibitors. These medicines help the immune system see and attack cancer cells better. Another type of treatment uses special medicines to stop cancer cells from making new blood vessels. Cancer cells need blood vessels to get food and grow, so stopping them is a good way to slow down the cancer.\\n\\nScientists are also looking at different types of kidney cancer that are not as common. These types don’t have as many treatment options yet. So, researchers are studying them to find new ways to help people with these kinds of cancer.\\n\\nThe goal is to find the best ways to help people with kidney cancer feel better and live longer. Scientists are learning more every day, and they hope to find even better treatments in the future!', 40677709: 'Hey there, little buddy! Let’s talk about some special medicine that helps people when they have a big ouchie called cancer. Scientists are trying to help people feel better faster.\\n\\nThere are special medicines called anti-PD-(L)1 agents. These medicines help our body’s soldiers, called cells, fight the bad stuff inside us. Right now, doctors use these medicines when the ouchie is really big. But scientists think using them earlier might be even better!\\n\\nThey did a study in a place called Switzerland. They wanted to see what happens if they use these medicines earlier for three types of ouchies: melanoma, renal cell carcinoma (RCC), and triple-negative breast cancer (TNBC). They looked at what happens over 10 years.\\n\\nGuess what they found? If doctors use these medicines earlier, fewer people get sick again. It’s like stopping the ouchie from coming back! They also found that fewer people need more medicine later, and more people stay healthy longer.\\n\\nThey think this could help a lot of people not just in Switzerland but everywhere. It could mean fewer ouchies and more happy days!\\n\\nSo, using these special medicines early could be a really good idea. It might help people stay healthy and happy much longer. Isn’t that cool?', 40677870: 'Alright, let\\'s talk about eyes and medicine in a simple way!\\n\\nSometimes, our eyes can have trouble if they don\\'t get enough blood. This can make it hard to see. Doctors call this problem \"ischemic optic neuropathy\" or ION for short. One kind of ION happens when the blood vessels in the eye get inflamed. This is called AAION.\\n\\nDoctors use special medicine called immune checkpoint inhibitors (ICIs) to help people with cancer. But sometimes, these medicines can cause eye problems. This story is about a 77-year-old man who had cancer and took ICIs. He started having trouble seeing. His doctors found out he had AAION because of the medicine.\\n\\nThe doctors used a special camera called an MRI to look at his eyes. They saw signs of AAION. They gave him medicine called steroids to help his eyes. The steroids helped a little, but his vision didn\\'t get all better.\\n\\nThis story teaches doctors to be careful and watch for eye problems when using ICIs. It\\'s important to find the right treatments to help people see better.\\n\\nSo, if you ever hear about someone having trouble seeing because of medicine, it might be a story like this. Doctors are working hard to find ways to help people see clearly and stay healthy!', 40677940: 'Let\\'s talk about a story from a hospital about a man who got better after a big surgery. This man, who was 56 years old, had a problem with his blood pressure. It was very high, and no medicine seemed to help. He also had headaches and felt his heart beating fast. The doctors felt something in his tummy and decided to take pictures inside using a special machine called a CT scan.\\n\\nThe pictures showed a big lump near his kidney, which is a part of the body that helps clean the blood. The doctors thought it might be a special kind of lump that makes people feel sick. But they couldn\\'t do more tests because they didn\\'t have the tools. So, they decided to do a big surgery to take out the kidney with the lump.\\n\\nAfter the surgery, the doctors looked at the lump under a microscope and found out it was a type of kidney lump called \"clear cell renal cell carcinoma.\" A special thing happened after the surgery: the man\\'s blood pressure got better! He didn\\'t need as much medicine anymore, and he felt good.\\n\\nThis story shows that sometimes, when someone has very high blood pressure that won\\'t go away, it might be because of a lump. Taking out the lump can help the person feel better. The doctors learned a lot from this case, and it helps them know what to look for in other people with the same problem.\\n\\nIn some places, doctors don\\'t have all the tools they need. But they can still help people by watching carefully and using the tools they do have. This story reminds doctors to think about lumps when they see someone with tricky blood pressure problems.', 40678334: \"This study is about a type of kidney cancer called renal cancer. Scientists want to find better ways to help people with this cancer. They are looking at a treatment called TRAIL, which tries to get cancer cells to stop growing. But sometimes, the cancer cells don't listen to TRAIL.\\n\\nThe scientists found something called chrysoeriol, which comes from plants. It can help TRAIL work better. When chrysoeriol and TRAIL are used together, they help the cancer cells start a process called apoptosis, which means the cells start to break down and go away.\\n\\nThe scientists did lots of tests in the lab. They mixed chrysoeriol and TRAIL with cancer cells and watched what happened. They saw that the cancer cells started to shrink and change shape. This means the cells were going through apoptosis and stopping their growth.\\n\\nThe scientists also looked at special parts inside the cells, like caspases and interleukins, which are like tiny messengers that tell the cells what to do. They found that when chrysoeriol and TRAIL were used together, these messengers helped the cancer cells go away.\\n\\nThe study showed that using chrysoeriol with TRAIL could be a good way to help people with this kind of cancer. But, scientists need to do more tests to make sure it is safe and works well in people, not just in the lab.\\n\\nIn the future, doctors might use this new combination treatment to help people with kidney cancer feel better. The scientists are excited because this could be a new way to help stop cancer cells from growing.\", 40678730: \"Hey there! Let's talk about something called endometrial cancer (EC). This is a type of cancer that happens in a part of the body called the uterus, which is important for having babies. Scientists are trying to learn more about how to help people with this kind of cancer.\\n\\nThere's a special plant called turmeric, and it has something inside it called curcumol. Curcumol might be able to help people with EC. Scientists are studying how it works. They found that curcumol might help by working with two special parts in our bodies. These parts have long names: Progesterone Receptor (PGR) and Ribosomal protein S6 kinase (RPS6KA1). We can just call them PGR and RPS6KA1.\\n\\nPGR and RPS6KA1 help control how cells grow and change. If curcumol can help these parts work better, it might help people with EC feel better and live longer.\\n\\nTo figure this out, scientists did lots of tests. They looked at how curcumol sticks to PGR and RPS6KA1. They found that curcumol fits really well with these parts, like a key in a lock. This means curcumol might help these parts do their job better.\\n\\nThe scientists also looked at how curcumol might help the body’s immune system, which is like a team of superheroes in our body that fights off bad stuff. They found that curcumol might help the immune system work better against EC.\\n\\nEven though these findings are exciting, scientists need to do more tests with animals and in the lab to make sure curcumol is safe and really works. This might take some time, but it could help make new medicines for EC in the future.\\n\\nSo, curcumol from turmeric might be a little helper in making medicine for EC. Scientists are working hard to learn more and help people feel better.\", 40681591: \"There's a kind of kidney sickness called ccRCC that scientists are trying to understand better. They are studying a special part of our cells named PDLIM7 to see how it affects this kidney problem.\\n\\nPDLIM7 is like a tiny worker in our cells that helps them do different jobs. Sometimes, this worker can do things that aren't good for the body. Scientists found out that when there's a lot of PDLIM7 in kidney cells, it might not be a good sign because it could make the sickness worse.\\n\\nTo learn more about PDLIM7, scientists did many tests. They looked at data from big computer databases and did experiments in labs with special tools. They found out that when they reduce the amount of PDLIM7 in kidney cells, the cells don't grow and spread as much. This is good because it means they might have found a way to help people with this kidney problem.\\n\\nThey also discovered that PDLIM7 could tell doctors how serious the kidney problem is in a person. Plus, they found that certain medicines work better when there's a lot of PDLIM7 in the cells. This could help doctors pick the best medicine for each person.\\n\\nIn simple words, PDLIM7 is an important part of our cells that can affect kidney health. By understanding it better, scientists hope to find new ways to help people with kidney problems. They think that maybe, in the future, they could use what they learn about PDLIM7 to make new medicines or to tell which treatment might work best for someone.\\n\", 40682515: \"Imagine your body is like a busy city with lots of workers (cells) doing different jobs. Sometimes, a few workers start doing things they shouldn't, and it can cause problems in the city. This is similar to what happens in a condition called monoclonal gammopathy of renal significance (MGRS). In MGRS, some cells make a special kind of protein that doesn't work right and can make the kidneys (a part of your body that cleans your blood) not work well.\\n\\nThere was a 74-year-old woman who had this problem. Her doctors found out that her kidneys were not working well because of the wrong proteins made by these cells. This woman also had a type of sickness called Waldenström macroglobulinemia, which is when even more wrong proteins are made and can make you feel very tired and weak.\\n\\nThe doctors decided to give her a special medicine called tirabrutinib. This medicine helped stop the cells from making the wrong proteins. After she started taking this medicine, her kidneys began to work better, and she stopped feeling so tired.\\n\\nThe doctors learned that it's really important to check the kidneys early if they think someone has MGRS, especially with the kind of protein this woman had, called IgM. Checking the kidneys early and starting the right medicine can really help make people feel better faster.\\n\\nSo, this story teaches us that if something in our body isn't working right, doctors can sometimes use special medicines to help fix the problem, just like fixing a part of the city that isn't working well. And it's always good to check things early, so small problems don't become big ones.\\n\", 40683967: \"Scientists are studying a special medicine called sunitinib that helps people with a kind of kidney cancer called ccRCC. But sometimes, this medicine stops working as well as it should. This happens because the cancer cells become resistant, meaning the medicine doesn't affect them much anymore.\\n\\nTo solve this problem, researchers did a big experiment using a special tool called CRISPR/Cas9, which helps change the genes in cells. They discovered that a part of the cell called PTGR2 might be helping the cancer cells resist the medicine. They thought that if they could stop PTGR2 from working, the medicine might start working better again.\\n\\nSo, they tried turning off PTGR2 in the cancer cells and found that sunitinib started working better at stopping the cancer cells from growing. They also looked deeper and found another part called KDM6A that works with PTGR2. When they turned off PTGR2, KDM6A also stopped working as much, which helped the medicine do its job better.\\n\\nThe scientists did more tests and found that if they could stop both PTGR2 and KDM6A from working, it helped the medicine work even better. They think this could be a new way to help people with this kind of kidney cancer when the usual medicine isn't working well on its own.\\n\\nThis research is important because it helps doctors understand better ways to help people when their cancer treatment isn't working as expected. It's like finding a new way to help the medicine fight the cancer better.\\n\", 40684090: 'Doctors and health experts in Italy are working on a special project to help take care of people\\'s health better. They are using a smart plan called \"\"risk stratification.\"\" This plan helps doctors understand who might need more help with their health soon.\\n\\nTo do this, a group of 24 health experts got together. They talked about what makes someone\\'s health get worse and how they can find this out from the health records. They agreed on what important things to look for, like if someone is very old, takes many medicines, or has tough illnesses like heart problems or cancer.\\n\\nThese experts used a special way of agreeing called the \"Delphi process.\" They asked each other questions and decided what health issues are very important to watch out for. They think about things like how old someone is, if they go to the hospital a lot, or if they feel very sad or stressed.\\n\\nThe idea is to make a tool that doctors can use to see who needs more care so they can help them before they get really sick. This tool will be tested to make sure it works well.\\n\\nThis work is part of a big effort to make sure everyone in Italy can have good health care, especially older people or those with many health problems. The experts hope this tool will help doctors take better care of their patients by understanding their health risks better.\\n', 40684099: \"A big study looked at heart health in people who have had cancer. Doctors know that sometimes, treatments for cancer can affect the heart, so they wanted to learn more about this.\\n\\nThey used information from a big survey that talked to lots of people across the United States from 2015 to 2020. They found about 1,700 people who had cancer, and they checked if these people also had heart problems like heart attacks or strokes.\\n\\nThe study found that about 21 out of every 100 people with cancer also had some kind of heart problem. It was a bit more common in men and in older people over 65. They also noticed that people who smoked, had high blood pressure, or diabetes were more likely to have heart problems.\\n\\nThe doctors also looked at what kinds of heart problems were most common. They found that blocked arteries and heart attacks were the most common heart issues for people with cancer.\\n\\nBy learning all this, doctors can do a better job of taking care of people's hearts when they are being treated for cancer. This is important because both cancer and heart disease are big reasons why people can get very sick or even die, so taking care of both at the same time can help people live longer and healthier lives.\\n\", 40684174: \"There is a special kind of research on a sickness called kidney cancer. The doctors and scientists are learning more about how this sickness can use a kind of fat called cholesterol to grow. They found out that a tiny loop-shaped thing inside our cells, called circABCA1, helps the cancer use more cholesterol. This tiny loop acts like a helper by making sure that the cancer cells can grab and keep more cholesterol from around them.\\n\\nThe doctors did a lot of tests in their labs. They looked at how much of this tiny loop is in the cancer cells. They found that if there's a lot of it, the cancer can do a better job of grabbing cholesterol. They also found out that this tiny loop can talk to another part inside the cells to help protect the cholesterol inside the cancer cells, so it doesn't get destroyed.\\n\\nThen, they did an experiment where they made less of this tiny loop in the cancer cells. They noticed that when they did this, the cancer cells couldn't hold onto as much cholesterol. This is good because it makes it harder for the cancer cells to grow.\\n\\nThey also learned that this tiny loop can affect the cells around the cancer too, making them behave in a way that helps the cancer. But, if they stop the loop from working, it can make the environment around the cancer less friendly for it to grow.\\n\\nSo, this research is really important because it helps doctors understand better ways to stop the cancer from getting the cholesterol it needs to grow. Maybe in the future, they can use this information to help people with this kind of kidney cancer feel better.\\n\", 40684203: 'Doctors and scientists use special math models to understand how medicine works in our bodies. These models help them figure out the best amount of medicine to give to people so that it works well and keeps them healthy.\\n\\nImagine your body is like a big mixing bowl. When you take medicine, it’s like adding a special liquid into the bowl that needs to mix well to work properly. The doctors use a special recipe to make sure the mix is just right. They check how long the medicine stays active and how much is needed to fight off the germs.\\n\\nRecently, some smart scientists made new recipes for a special kind of medicine given through a tiny tube into the veins, which is a common way to give medicine in hospitals. They found a better way to figure out how much medicine someone needs in a day and how long it should stay in the body to work best.\\n\\nThey tested their new recipes on pretend patients in a computer to make sure they work well. The new recipes were better than the old ones because they could more accurately decide the right amount of medicine. This is really helpful for doctors when they need to give the right amount of medicine to their patients to make them feel better faster.\\n\\nUsing these new recipes, doctors can make sure the medicine mixes just right in the big mixing bowl of our bodies, helping everyone get the right treatment and feel better sooner!\\n', 40686509: \"Imagine a man, 56 years old, who felt very weak and had trouble moving his legs. He went to the hospital three times because he was not feeling well. The doctors found out that he didn't have enough potassium, a very important mineral that helps our muscles work right.\\n\\nThe doctors did some tests and saw a big lump in his kidney. They thought it might be a kind of lump called RCC, which usually needs to be removed because it can be troublesome. So, they decided to take out his kidney to help him feel better.\\n\\nAfter the surgery, the doctors looked closely at the lump and found out it was not RCC but another kind called AML. This type is usually not troublesome and doesn't need the whole kidney removed. The good news is, after the surgery, the man's potassium level went back to normal, and he felt strong again.\\n\\nThe doctors learned a lot from this. They found out that sometimes AML can look like RCC and make it hard to decide the best way to help. If they had known it was AML from the start, they might have chosen a simpler way to help him.\\n\\nThis story helps doctors think about how to look for and understand different kinds of lumps in the kidney. It's important because it helps them make the best choices for their patients.\\n\", 40687426: \"A 73-year-old woman  had a few health challenges. She had a kidney transplant a long time ago and needed to take special medicines to keep her body from rejecting the new kidney. These medicines made her body less able to fight off germs and viruses.\\n\\nThis woman caught a virus called HPV, which is very common and can be passed from one person to another. Usually, our bodies can fight off HPV, but because the woman's body was not as strong in fighting germs due to her medicines, the virus stayed in her body. Over time, this virus caused her to develop three different kinds of lumps in her lower body areas, which doctors call tumors.\\n\\nDoctors found out that these lumps were there because of the HPV virus, and they needed to treat them to help the woman feel better. They did some surgeries to remove the lumps and also gave her some other treatments to make sure all the unhealthy cells were gone.\\n\\nThis story teaches us that it's really important for doctors to keep a close eye on people who have had transplants or who take medicines that affect their immune systems. It also shows us how helpful vaccines can be. Just like we get shots to protect us from the flu or measles, there's a shot for HPV which can help stop these kinds of health problems from happening.\\n\\nThe woman continued to visit her doctors regularly to make sure she stayed healthy, and they kept checking to make sure the virus was not causing any more problems. This helps us understand how doctors and patients work together to keep an eye on health, especially when someone has a condition that makes them more likely to get sick.\\n\", 40687644: \"Doctors and scientists are always working to learn more about how to help people who get sick with a type of sickness called clear cell renal cell carcinoma, which is a kind of kidney sickness. They use information from big databases to understand better who might get very sick and who might get a little sick from this type of kidney problem.\\n\\nIn this study, the researchers looked at something called EMT-association genes. These are special instructions in our body that can affect how the kidney sickness behaves and spreads. The scientists used computers to find out which of these instructions are really important and can tell doctors how sick someone might get.\\n\\nThey found 12 important instructions and used them to make a special scoring system. This system helps doctors figure out if a person with this kidney sickness might get very sick or not so sick. They tested this system with information from many people and found out it works pretty well.\\n\\nThe scoring system also helps doctors understand how the sickness affects the body's defense system, which fights off sicknesses. The system shows that some people might need different medicines based on how their body's defense system is working.\\n\\nIn simple words, this study helps doctors by giving them a tool to know how serious the kidney sickness might be in someone and how they might respond to different treatments. This means doctors can choose the best medicine for each person, which is really great for helping everyone feel better.\\n\", 40687659: \"A group of scientists did a study to see if a special type of water called hydrogen-rich water (HRW) could help with a health problem called diabetes. Diabetes is when a person has high sugar levels in their blood, which can cause problems in different parts of the body.\\n\\nThe scientists used rats to find out how this special water works. They gave some rats the HRW to drink for about six months and watched what happened. They found out that the rats drinking the HRW had better sugar levels in their blood and were healthier overall. They also noticed that these rats had fewer problems like eye issues and less weight loss compared to rats that didn't drink the special water.\\n\\nThe scientists also looked inside the rats' bodies to see what was happening. They found changes in the rats' genes, which are like tiny instructions inside every animal and person that tell the body how to work. These changes helped the rats use their food better and stay healthier.\\n\\nFrom this study, the scientists learned that drinking HRW might be a good way to help people with diabetes feel better and have fewer health problems. They think this special water could change how the body handles sugar and other important stuff inside.\\n\\nIn simple words, the study shows that this special water could be a helpful drink for people with diabetes, helping them to stay healthier and happier.\\n\", 40687664: \"Doctors and scientists have been working hard to help people with a serious kidney problem called stage IV renal cancer. They looked at a lot of information from a big database in Japan to see if new medicines are helping people get better.\\n\\nThey found that over time, as new medicines were introduced, people with this kidney problem started doing better. They divided their study into three time periods to see the changes clearly. The first period was until March 2008, the second from April 2008 to July 2016, and the third from August 2016 onwards.\\n\\nIn their study, they found that people who could have surgery to remove part of the kidney did really well, especially in the later periods when new medicines were available. However, for people whose cancer had spread too much and couldn't have surgery, the new medicines didn't make as much of a difference.\\n\\nThey also noticed that the new medicines seemed to help more if the cancer hadn't spread too far. For those whose cancer had spread a lot, even the new medicines couldn't help as much as they hoped.\\n\\nThe doctors and scientists keep looking at the information to learn how to help people with this kidney problem better. They hope that with more research, they can find even better ways to use medicines and surgery to help everyone who has this illness.\\n\", 40687984: \"Imagine a small tube called a catheter that doctors use to help people pee when they can't do it by themselves. Sometimes, this tube can get tangled up like your shoelaces. This doesn't happen a lot, but it did happen to a 79-year-old man. He was already in the hospital because he was very confused and his kidneys were not working right.\\n\\nThe doctors found out that the pee tube had tied itself in a knot around another tube, called a stent, that was supposed to help his kidney work better. This was a big problem because it made the man even sicker.\\n\\nThe doctors tried to untangle the knot with some special tools while looking at it on a screen, kind of like a video game, but that didn't work. So, they decided to do a small surgery. They put the man to sleep, and then they used a special tool called a stone crusher to break the knot into small pieces and take it out. This tool is usually used to break up kidney stones, but it worked well for the knot too!\\n\\nAfter the surgery, the doctors put in a new stent and a different tube to help the man pee without any problems. They checked everything with a special camera to make sure it was all okay.\\n\\nThis story is very rare, and not many doctors see this happen. But it's important for doctors to know about it, so if it does happen, they can fix it quickly. The doctors learned that using the stone crusher is a good way to solve this tricky problem. They also learned to always check if the tube is tangled before pulling it, to keep this from happening.\\n\", 40688377: \"Imagine your body is like a house with lots of pipes that carry water everywhere it's needed. Now, think about what happens if one of those pipes gets squeezed and the water can't flow well. That's a bit like what happens in a part of the body called the kidney when there's too much pressure inside it.\\n\\nDoctors and scientists did a special study using big pigs because their kidneys are a lot like ours. They wanted to see what happens when there's too much pressure in the kidneys. To do this, they used a special camera that can see things we can't, like how much oxygen is in the kidney and how well blood is flowing through it.\\n\\nThey found out that when there is too much pressure, the kidney doesn't get enough blood and oxygen, especially at the ends of the kidney. This isn't good because the kidney needs oxygen and blood to stay healthy.\\n\\nThe camera showed them pictures that helped them understand how the kidney was doing. This is really helpful for doctors when they are doing surgery or treating people with kidney problems. They learned that they need to keep an eye on the pressure to make sure the kidneys are okay.\\n\\nThis study helps doctors take better care of people's kidneys by showing them how important it is to not let the pressure get too high. It's like making sure the water pipes in your house aren't too tight so everything keeps working right.\\n\", 40689304: \"A 60-year-old man from China went to the hospital because he had a health check and found out something was off with his blood tests. He had too much of a substance called PTH and not enough of another called phosphorus, but his calcium was okay. This was unusual because usually, if someone has too much PTH, their calcium is also high.\\n\\nDoctors kept an eye on him and two years later, they found out he had prostate cancer, which is a sickness in the prostate, a small part of a man's body near the bladder. They checked his bones and found that the cancer hadn't spread there, which was good news.\\n\\nThe doctors gave him treatment for his prostate cancer. After the treatment, the levels of PTH and phosphorus in his blood went back to normal. This was interesting because usually, PTH comes from a different place near the throat called the parathyroid glands, not the prostate.\\n\\nThe doctors did more tests and found out that his prostate cancer was making the PTH, which is very rare. They think this might be why his PTH was high. After they treated his cancer, his PTH levels got better because they were treating the real cause, not just the symptoms.\\n\\nThis story helps doctors understand that sometimes, when something unusual shows up in tests, it might be because of a rare condition. It's like being a detective and finding clues to solve a mystery. This helps doctors take better care of people by understanding more about how our bodies work.\\n\", 40691606: \"There's a special protein in our bodies called FH that helps turn the food we eat into energy. Sometimes, changes in the FH gene can cause problems, like making some people more likely to get sick with a kind of kidney problem. Scientists wanted to understand more about how different changes in the FH gene affect people's health.\\n\\nSo, they looked at lots of studies from different places and put all the information together to see what they could find. They found out that some changes in the FH gene might make it more likely for people to get sick than other changes. They think this information could help doctors watch out for and take care of people who might get this kidney problem because of their FH gene.\\n\\nThey also found that where people live, like in North America or Europe, might make a difference in how likely they are to get sick. And they noticed that the way scientists look for these gene changes can affect what they find.\\n\\nIn the end, the scientists learned a lot, but they still think they need to do more studies to understand better. They hope that knowing more about the FH gene will help doctors take better care of people who might get sick because of changes in this gene.\\n\", 40692681: \"Doctors did a study to learn how the tiny bugs in our guts, called gut microbiota, affect older men's health when they take many medicines for different health problems. They looked at 33 older men, around 94 years old, who took lots of medicines to see if the kinds of bugs in their guts could tell us something about their health after 18 months.\\n\\nThey found out that certain scores that measure health problems and levels of a substance called creatinine in the blood could help guess how healthy someone would be. They also noticed that different types of tiny bugs in the gut were more common in men who stayed healthier compared to those who didn't.\\n\\nThe study also talked about how as people get older, they often have to take more medicines and might have more than one health problem. This can make their health complicated because all these medicines can affect each other and the body in many ways.\\n\\nThe tiny bugs in our guts are very important because they help us digest food, protect us from getting sick, and even affect our mood. But as we get older, these bugs can change, and not always in good ways. This study tried to see if these changes could help predict health problems in older men taking many medicines.\\n\\nIn simple words, the study tried to find clues in the gut bugs that could help doctors take better care of older people who need a lot of medicines. They learned some interesting things that might help in the future, but they need to do more studies to be sure.\\n\", 40692785: \"Imagine your body has a team of tiny helpers called inflammasomes. They are like little security guards that keep you safe from germs and other harmful stuff. When these guards spot danger, they sound an alarm that helps your body protect itself. This alarm can help stop bad things like a cold or flu.\\n\\nBut sometimes, these security guards can get a bit too excited. If they keep sounding the alarm too much, it can cause problems, like making parts of your body swell up. This swelling is called inflammation. In some cases, if there's too much inflammation, it can make it easier for illnesses to happen, even things like cancer, which is when some cells in your body grow too fast and cause trouble.\\n\\nThe inflammasomes work by spotting the bad stuff and then sending signals to start the inflammation. They look for two types of danger signals. One type comes from germs, and the other type comes from your own body when it's hurt. When they find these signals, they gather and set off a reaction that helps clear out the bad stuff but can also cause swelling.\\n\\nSometimes, doctors use treatments like radiation, which is a special kind of energy, to help get rid of cancer. But this energy can also make the inflammasomes very active and cause more swelling, which can make things a bit tricky.\\n\\nResearchers are very interested in these inflammasomes because they can both help and cause problems. They are trying to figure out how to keep the good effects, like getting rid of germs, while stopping the bad effects, like too much swelling. By understanding more about how these tiny security guards work, doctors can better help people stay healthy and treat illnesses like cancer more effectively.\\n\", 40692977: \"Doctors and scientists are learning how being very overweight affects our health. They use a special method called Mendelian randomization (MR) to understand if being overweight causes health problems. MR looks at our genes, which are like tiny instructions inside our bodies that make us who we are.\\n\\nSome people have certain gene changes that make them more likely to be overweight. Scientists use MR to see if these gene changes also make people more likely to get sick with things like heart problems or diabetes. They found that being very overweight can indeed make these health problems more likely.\\n\\nScientists also discovered that genes can tell us why some people get very overweight. They found some special genes that affect how we use energy from food, how hungry we feel, and how our bodies handle sugar and fat.\\n\\nUnderstanding these genes helps doctors think of new ways to help people who are very overweight. They can make special medicines that target these genes to help people feel less hungry or use their body's sugar better.\\n\\nOverall, MR helps scientists learn a lot about being overweight and health. It shows that our genes play a big part in our health, especially how our weight affects our chances of getting sick. This helps everyone know more about how to stay healthy.\\n\"}\n"
     ]
    }
   ],
   "source": [
    "#Read csv columns and assign columns\n",
    "import pandas as pd\n",
    "df=pd.read_csv(\"/Users/t2yu/Downloads/Nov12 Updated Data - renal_cancer_20251112_135526.csv\")\n",
    "pmids = df['pmid'].tolist()\n",
    "full_text_summaries = df.set_index('pmid')['Full_Text_Summary'].to_dict()\n",
    "abstract_summaries = df.set_index('pmid')['Abstract_Summary'].to_dict()\n",
    "print(full_text_summaries)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2677839a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PMID 40665378: Autophagy inhibition potentiates anti-cancer activity of Sunitinib in kidney cancer cells.\n",
      "Authors: Patergnani S, Zampieri P, Bianchi N, Ippolito C, Gafà R, Lanza G, Wieckowski MR, Pinton P, Aguiari G\n",
      "Journal: Biol Direct\n",
      "---\n",
      "PMID 40665456: The tumour sink effect on 68Ga-PSMA-PET/CT in metastatic castration-resistant prostate cancer and its implications for PSMA-RPT: a sub-analysis of the 3TMPO study.\n",
      "Authors: Zamanian A, Rousseau É, Buteau FA, Arsenault F, Beaulieu A, April G, Juneau D, Plouznikoff N, Turcotte ÉE, Allard C, Richard PO, Saad F, Guérin B, Pouliot F, Beauregard JM\n",
      "Journal: Cancer Imaging\n",
      "---\n",
      "PMID 40666060: Catastrophic antiphospholipid syndrome in a young patient: a case report of severe multi-organ and cardiovascular involvement.\n",
      "Authors: Desamour P, Alshwayyat S, Alvarez L, Almasri N, Hanifa H\n",
      "Journal: Oxf Med Case Reports\n",
      "---\n",
      "PMID 40667694: A systematic review of the latest progress of drug resistance and clinical application of immunotherapy in renal cell carcinoma in the past five years based on bibliometrics.\n",
      "Authors: Wu Y, Liang J, Liang Y, Ma Y\n",
      "Journal: Hum Vaccin Immunother\n",
      "---\n",
      "PMID 40669903: Patient-reported outcome measures for fatigue in patients with chronic kidney disease: a systematic review.\n",
      "Authors: Hughes A, Ju A, Cazzolli R, Howell M, Guha C, Levin A, Manera K, Teixeira-Pinto A, Torrisi LG, Wheeler DC, Wong G, Wu R, Jaure A\n",
      "Journal: BMJ Open\n",
      "---\n",
      "PMID 40671833: Effect of liver biopsy size on MASLD fibrosis assessment by second-harmonic generation/two-photon excitation fluorescence microscopy.\n",
      "Authors: Field DT, Ren Y, Akbary K, Chng E, Tai D, Naoumov NV, Kleiner DE, Fallowfield JA, Kendall TJ, Sanyal AJ\n",
      "Journal: JHEP Rep\n",
      "---\n",
      "PMID 40671868: Early clinical experience with the Carina robotic platform in urologic surgery.\n",
      "Authors: Wang H, Cao D, Moschovas MC, Wu J, Wang B, Chang K, Wang H, Zhu Y, Dai B, Zhu Y, Zhang H, Patel VR, Ye D\n",
      "Journal: BJUI Compass\n",
      "---\n",
      "PMID 40672051: Quantifying diagnostic intervals and routes to diagnosis for children and young people with cancer in the UK (Childhood Cancer Diagnosis study, CCD): a population-based observational study.\n",
      "Authors: Shanmugavadivel D, Liu JF, Ritzmann TA, Ball-Gamble A, Polanco A, Ranasinghe N, Vedhara K, Ojha S, Walker D\n",
      "Journal: Lancet Reg Health Eur\n",
      "---\n",
      "PMID 40673289: Immune Checkpoint Inhibitor-Induced Myositis, Myocarditis, and Myasthenia Gravis Overlap Syndrome Following Nivolumab and Ipilimumab Combination Therapy in a 60-Year-Old Male With Metastatic Renal Cell Carcinoma.\n",
      "Authors: Ismail U\n",
      "Journal: Cureus\n",
      "---\n",
      "PMID 40673609: Identification and Validation of a Prognostic Model Based on Tumour Necrosis Factor-Related mRNAs for Kidney Renal Clear Cell Carcinoma.\n",
      "Authors: Hu Z, Zhou Y, Xie L, Zhang S, Liu Y, Zhang W, Huang T\n",
      "Journal: J Cell Mol Med\n",
      "---\n",
      "PMID 40673839: Roadblocks of Urinary EV Biomarkers: Moving Toward the Clinic.\n",
      "Authors: Droste M, Puhka M, van Royen ME, Ng MSY, Blijdorp C, Alvarez-Llamas G, Borràs FE, Büscher AK, Bussolati B, Dear JW, Falcón-Pérez JM, Giebel B, Grange C, Hoorn EJ, Leivo J, Lenassi M, Llorente A, Lucien F, Mertens I, Mischak H, Pink D, Tertel T, Tiwari S, Di Vizio D, Yuen PST, Zarovni N, Jenster G, Burger D, Martens-Uzunova ES, Erdbrügger U\n",
      "Journal: J Extracell Vesicles\n",
      "---\n",
      "PMID 40677495: Synchronous Bladder Metastasis From Chromophobe Renal Cell Carcinoma: A Report of a Rare Case.\n",
      "Authors: Romero Uribe JL, Cornejo Davila V, Cruz Ruiz J\n",
      "Journal: Cureus\n",
      "---\n",
      "PMID 40677502: From Cholesterol Control to Critical Illness: A Case of Statin-Associated Rhabdomyolysis With Concurrent Acute Kidney Injury and Hepatotoxicity.\n",
      "Authors: Habeb B, Khair S, Fowler S\n",
      "Journal: Cureus\n",
      "---\n",
      "PMID 40677703: Advances in immunotherapy and targeted therapy for advanced clear-cell renal cell carcinoma: current strategies and future directions.\n",
      "Authors: Wang B, Xiang Y, Liu X, Pan X, Peng L, Ma J\n",
      "Journal: Front Immunol\n",
      "---\n",
      "PMID 40677709: Health impact of using anti-PD-(L)1 agents to treat early-stage cancers in Switzerland: a modeling study.\n",
      "Authors: Favre-Bulle A, Stanković M, Mantaian T, Frei C, Schaefer S, Gabriel SS, Sönmez D, Aguiar-Ibáñez R\n",
      "Journal: Front Immunol\n",
      "---\n",
      "PMID 40677870: Arteritic anterior ischemic optic neuropathy secondary to immune checkpoint inhibitors: A case report.\n",
      "Authors: Alam Z, Siddiqui TA, Syed MU, Kottamasu PK, Nag T, Rahman A\n",
      "Journal: Radiol Case Rep\n",
      "---\n",
      "PMID 40677940: Resolution of Refractory Hypertension Following Radical Nephrectomy for Renal Cell Carcinoma: A Case Report from Somalia in Resource Limit Setting.\n",
      "Authors: Ali RM, Omar AA, Ali IA\n",
      "Journal: Cancer Manag Res\n",
      "---\n",
      "PMID 40678334: Chrysoeriol Chemosensitized Renal Cell Carcinoma (RCC) Cell Lines to Trigger TRAIL Induced Apoptosis In Vitro by Up-Regulating Pro-Apoptotic and Down-Regulating Anti-Apoptotic Genetic Factors.\n",
      "Authors: Batool Z, Wu Q, Kamal MA, Weng G, Shen B\n",
      "Journal: Food Sci Nutr\n",
      "---\n",
      "PMID 40678730: Multi-omics analysis of the anti-cancer effects of curcumol in endometrial carcinoma.\n",
      "Authors: Xie J, Zhu Y, Bai H, Song Z, Qiu S, Xie Y, Wan S\n",
      "Journal: Front Pharmacol\n",
      "---\n",
      "PMID 40681591: The effect of PDLIM7 on cell proliferation, migration, and drug sensitivity in clear cell renal cell carcinoma.\n",
      "Authors: Shen C, Zhuo J, Wang J\n",
      "Journal: Sci Rep\n",
      "---\n",
      "PMID 40682515: Progression of IgM Monoclonal Gammopathy of Renal Significance (MGRS) to Symptomatic Waldenström Macroglobulinemia: A Case Report.\n",
      "Authors: Ito K, Shimoyamada H, Hirano K, Sekiguchi N\n",
      "Journal: Cancer Rep (Hoboken)\n",
      "---\n",
      "PMID 40683967: Genome-wide CRISPR/Cas9 screening identifies PTGR2 as a potential therapeutic target for sunitinib resistance in clear cell renal cell carcinoma.\n",
      "Authors: Chang W, Zhang B, Yang S, Zhang X, Zhang L, Zhang S, Lu J, Wang W, Shang P, Yue Z\n",
      "Journal: Sci Rep\n",
      "---\n",
      "PMID 40684090: Stratifying the population based on health risk: identification of patient key health risk factors through consensus techniques.\n",
      "Authors: Castagna C, Huff A, Douglas A, Garofano M, Fabi M, Hass R, Maio V\n",
      "Journal: BMC Prim Care\n",
      "---\n",
      "PMID 40684099: Prevalence and clinical risk factors of cardiovascular disease in patients with cancer: a cross-sectional study.\n",
      "Authors: Jiang C, Chen YX, Geng QS, Liu AB\n",
      "Journal: BMC Cardiovasc Disord\n",
      "---\n",
      "PMID 40684174: CircABCA1 promotes ccRCC by reprogramming cholesterol metabolism and facilitating M2 macrophage polarization through IGF2BP3-mediated stabilization of SCARB1 mRNA.\n",
      "Authors: Ning H, Jiang Y, Li B, Ren J, Wang C, Wei L, Li L, Ran A, Li Z, Li J, Li W, Wang Y, Xiao B\n",
      "Journal: Mol Cancer\n",
      "---\n",
      "PMID 40684203: Modeling of pharmacokinetic/pharmacodynamic parameters in regular intermittent intravenous infusion and translational application of the models in personalized antibiotics dosing.\n",
      "Authors: Song X\n",
      "Journal: J Transl Med\n",
      "---\n",
      "PMID 40686509: Resolution of hypokalemia following nephrectomy for a presumed renal cell carcinoma revealing a large angiomyolipoma: A case report.\n",
      "Authors: Khajeh F, Zeighami S, Ariafar A, Bazargani Z, Hakimelahi H\n",
      "Journal: Urol Case Rep\n",
      "---\n",
      "PMID 40687426: Human papillomavirus-associated anal cancer, vulvar verrucous cancer, and cervical cancer in a post-renal transplant patient-a case report.\n",
      "Authors: Feng X, Shen QB, Yuan K, Mao Y, Wang X, Min AP\n",
      "Journal: Front Oncol\n",
      "---\n",
      "PMID 40687644: Epithelial-mesenchymal transition classification based on machine learning for predicting prognosis and treatment response in clear cell renal cell carcinoma.\n",
      "Authors: Zhu G, Tang R, Tang T, Wei X, Tan C\n",
      "Journal: Transl Androl Urol\n",
      "---\n",
      "PMID 40687659: Long-term administration of hydrogen-rich water enhances quality of life in diabetic rats and reduces renal neoplasm incidence through modulation of inflammatory and metabolic pathways.\n",
      "Authors: Chen Q, Yu S, Pan Y, Tang H, Wang Y, Feng N, Wu H, Yu G\n",
      "Journal: Transl Androl Urol\n",
      "---\n",
      "PMID 40687664: Improvements in the prognosis of patients with stage IV renal cancer according to the era of therapeutic agents: an analysis of National Cancer Registry data.\n",
      "Authors: Watari S, Ichikawa T, Takao S, Kurihara Y, Matsumi A, Tokunaga M, Kubota R, Kusumi N, Tsushima T\n",
      "Journal: Transl Androl Urol\n",
      "---\n",
      "PMID 40687984: Knotting of a Urinary Catheter and Ureteric Stent: A Unique Complication and Management Solution.\n",
      "Authors: Logan H, Lockhart K, Chong P\n",
      "Journal: Case Rep Urol\n",
      "---\n",
      "PMID 40688377: Hyperspectral Imaging Accurately Detects Renal Malperfusion Due to High Intrarenal Pressure.\n",
      "Authors: Egen L, Hommel M, Haney CM, Özdemir B, Knoedler S, Sellner J, Seidlitz S, Dietrich M, Salg GA, Nickel F, Maier-Hein L, Michel MS, Studier-Fischer A, Kowalewski KF\n",
      "Journal: Eur Urol Open Sci\n",
      "---\n",
      "PMID 40689304: Ectopic Parathyroid Hormone Expression in Nonmetastatic Prostate Cancer.\n",
      "Authors: Rama Chandran S, Tan YB\n",
      "Journal: JCEM Case Rep\n",
      "---\n",
      "PMID 40691606: Association between fumarate hydratase variant subtypes and the risk of HLRCC-associated renal cell carcinoma: systematic review and meta-analysis.\n",
      "Authors: Wang H, Rehim S, Wang H\n",
      "Journal: Hum Genomics\n",
      "---\n",
      "PMID 40692681: Investigating potential links between gut microbiome, clinical parameters, and mortality in long-living male patients receiving multi-drug therapy.\n",
      "Authors: Wu Q, Pan H, Zeng X, Xiong F, Lv Y, Jiang Z, Li J, Wang L, Wang X\n",
      "Journal: Front Cell Infect Microbiol\n",
      "---\n",
      "PMID 40692785: Regulation of cancer by inflammasomes: from inflammation to tumorigenesis.\n",
      "Authors: Malvankar S, Jaiswal P, Bhat PP, Mehto S\n",
      "Journal: Front Immunol\n",
      "---\n",
      "PMID 40692977: Advances in Mendelian Randomization Studies of Obesity Over the Past Decade: Uncovering Key Genetic Mechanisms.\n",
      "Authors: Lu X, Ji L, Chen D, Lian X, Yuan M\n",
      "Journal: Diabetes Metab Syndr Obes\n",
      "---\n",
      "Retrieved PMIDs: [40665378, 40665456, 40666060, 40667694, 40669903, 40671833, 40671868, 40672051, 40673289, 40673609, 40673839, 40677495, 40677502, 40677703, 40677709, 40677870, 40677940, 40678334, 40678730, 40681591, 40682515, 40683967, 40684090, 40684099, 40684174, 40684203, 40686509, 40687426, 40687644, 40687659, 40687664, 40687984, 40688377, 40689304, 40691606, 40692681, 40692785, 40692977]\n",
      "Number of PMIDs retrieved: 38\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/4s/6g39d69s7m35v6gzfsln6thr0000gp/T/ipykernel_16357/1355655258.py:106: DeprecationWarning: LangchainLLMWrapper is deprecated and will be removed in a future version. Use the modern LLM providers instead: from ragas.llms.base import llm_factory; llm = llm_factory('gpt-4o-mini') or from ragas.llms.base import instructor_llm_factory; llm = instructor_llm_factory('openai', client=openai_client)\n",
      "  evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model=\"gpt-4o\"))\n",
      "Evaluating:   0%|          | 0/38 [00:00<?, ?it/s]2025-11-12 15:09:46 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:46 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:46 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:47 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:47 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:47 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:48 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:48 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:49 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:49 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:49 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:50 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:50 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:50 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:50 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:51 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:54 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:   3%|▎         | 1/38 [00:11<06:50, 11.10s/it]2025-11-12 15:09:54 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:   5%|▌         | 2/38 [00:11<02:59,  4.97s/it]2025-11-12 15:09:56 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:   8%|▊         | 3/38 [00:13<01:59,  3.41s/it]2025-11-12 15:09:56 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  11%|█         | 4/38 [00:13<01:13,  2.17s/it]2025-11-12 15:09:57 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  13%|█▎        | 5/38 [00:14<00:52,  1.59s/it]2025-11-12 15:09:57 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  16%|█▌        | 6/38 [00:14<00:36,  1.14s/it]2025-11-12 15:09:57 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  18%|█▊        | 7/38 [00:14<00:26,  1.18it/s]2025-11-12 15:09:58 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:58 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  21%|██        | 8/38 [00:15<00:24,  1.24it/s]2025-11-12 15:09:59 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:09:59 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  24%|██▎       | 9/38 [00:16<00:28,  1.03it/s]2025-11-12 15:10:00 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:00 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:00 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:00 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:00 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  26%|██▋       | 10/38 [00:17<00:28,  1.02s/it]2025-11-12 15:10:01 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  29%|██▉       | 11/38 [00:18<00:26,  1.01it/s]2025-11-12 15:10:02 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:02 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:02 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:03 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:03 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  34%|███▍      | 13/38 [00:20<00:21,  1.17it/s]2025-11-12 15:10:03 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:04 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  39%|███▉      | 15/38 [00:20<00:14,  1.64it/s]2025-11-12 15:10:05 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  42%|████▏     | 16/38 [00:22<00:19,  1.12it/s]2025-11-12 15:10:06 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:07 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:07 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:07 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  45%|████▍     | 17/38 [00:24<00:24,  1.18s/it]2025-11-12 15:10:08 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  47%|████▋     | 18/38 [00:24<00:18,  1.07it/s]2025-11-12 15:10:08 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:08 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:09 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:10 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  50%|█████     | 19/38 [00:27<00:24,  1.29s/it]2025-11-12 15:10:10 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:11 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  53%|█████▎    | 20/38 [00:28<00:20,  1.16s/it]2025-11-12 15:10:12 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:12 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:12 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:12 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  58%|█████▊    | 22/38 [00:29<00:16,  1.01s/it]2025-11-12 15:10:13 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:14 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  61%|██████    | 23/38 [00:30<00:15,  1.04s/it]2025-11-12 15:10:14 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:16 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  63%|██████▎   | 24/38 [00:33<00:20,  1.44s/it]2025-11-12 15:10:16 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  66%|██████▌   | 25/38 [00:33<00:14,  1.13s/it]2025-11-12 15:10:17 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:17 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  68%|██████▊   | 26/38 [00:34<00:11,  1.02it/s]2025-11-12 15:10:19 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  71%|███████   | 27/38 [00:36<00:12,  1.18s/it]2025-11-12 15:10:19 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  74%|███████▎  | 28/38 [00:36<00:09,  1.09it/s]2025-11-12 15:10:20 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  76%|███████▋  | 29/38 [00:36<00:07,  1.18it/s]2025-11-12 15:10:20 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  79%|███████▉  | 30/38 [00:37<00:06,  1.33it/s]2025-11-12 15:10:21 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  82%|████████▏ | 31/38 [00:38<00:04,  1.49it/s]2025-11-12 15:10:21 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  84%|████████▍ | 32/38 [00:38<00:04,  1.46it/s]2025-11-12 15:10:22 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  87%|████████▋ | 33/38 [00:38<00:02,  1.80it/s]2025-11-12 15:10:23 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:25 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  89%|████████▉ | 34/38 [00:41<00:05,  1.27s/it]2025-11-12 15:10:25 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  92%|█████████▏| 35/38 [00:42<00:02,  1.03it/s]2025-11-12 15:10:28 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  95%|█████████▍| 36/38 [00:45<00:03,  1.62s/it]2025-11-12 15:10:31 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  97%|█████████▋| 37/38 [00:48<00:02,  2.02s/it]2025-11-12 15:10:38 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating: 100%|██████████| 38/38 [00:55<00:00,  1.45s/it]\n",
      "Evaluating:   0%|          | 0/38 [00:00<?, ?it/s]2025-11-12 15:10:42 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:42 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:42 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:42 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:43 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:43 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:43 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:43 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:10:44 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:10 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:   3%|▎         | 1/38 [00:31<19:18, 31.31s/it]2025-11-12 15:11:25 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:   5%|▌         | 2/38 [00:46<13:09, 21.92s/it]2025-11-12 15:11:25 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:26 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:   8%|▊         | 3/38 [00:47<07:09, 12.27s/it]2025-11-12 15:11:27 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  11%|█         | 4/38 [00:49<04:36,  8.12s/it]2025-11-12 15:11:30 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:31 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  13%|█▎        | 5/38 [00:52<03:30,  6.36s/it]2025-11-12 15:11:32 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:32 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  16%|█▌        | 6/38 [00:54<02:32,  4.78s/it]2025-11-12 15:11:33 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  18%|█▊        | 7/38 [00:54<01:42,  3.30s/it]2025-11-12 15:11:33 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:35 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  21%|██        | 8/38 [00:56<01:27,  2.91s/it]2025-11-12 15:11:36 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  24%|██▎       | 9/38 [00:57<01:10,  2.42s/it]2025-11-12 15:11:37 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  26%|██▋       | 10/38 [00:58<00:52,  1.86s/it]2025-11-12 15:11:37 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:38 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:39 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:39 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:41 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:42 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:46 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  29%|██▉       | 11/38 [01:07<01:52,  4.17s/it]2025-11-12 15:11:47 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  32%|███▏      | 12/38 [01:08<01:20,  3.10s/it]2025-11-12 15:11:49 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  34%|███▍      | 13/38 [01:11<01:13,  2.94s/it]2025-11-12 15:11:51 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:51 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  37%|███▋      | 14/38 [01:12<01:00,  2.53s/it]2025-11-12 15:11:52 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:54 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:11:55 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:01 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  39%|███▉      | 15/38 [01:22<01:51,  4.86s/it]2025-11-12 15:12:02 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  42%|████▏     | 16/38 [01:23<01:16,  3.49s/it]2025-11-12 15:12:04 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  45%|████▍     | 17/38 [01:25<01:04,  3.09s/it]2025-11-12 15:12:04 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:06 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:07 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:10 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  47%|████▋     | 18/38 [01:31<01:20,  4.04s/it]2025-11-12 15:12:14 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:17 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  50%|█████     | 19/38 [01:39<01:36,  5.07s/it]2025-11-12 15:12:19 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  53%|█████▎    | 20/38 [01:41<01:14,  4.16s/it]2025-11-12 15:12:21 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  55%|█████▌    | 21/38 [01:42<00:56,  3.33s/it]2025-11-12 15:12:22 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  58%|█████▊    | 22/38 [01:43<00:43,  2.74s/it]2025-11-12 15:12:23 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  61%|██████    | 23/38 [01:44<00:30,  2.04s/it]2025-11-12 15:12:23 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:24 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  63%|██████▎   | 24/38 [01:45<00:26,  1.92s/it]2025-11-12 15:12:25 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  66%|██████▌   | 25/38 [01:47<00:21,  1.69s/it]2025-11-12 15:12:29 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:30 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:32 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  68%|██████▊   | 26/38 [01:53<00:37,  3.16s/it]2025-11-12 15:12:33 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  71%|███████   | 27/38 [01:54<00:27,  2.49s/it]2025-11-12 15:12:33 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-12 15:12:35 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  74%|███████▎  | 28/38 [01:56<00:24,  2.40s/it]2025-11-12 15:12:36 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  76%|███████▋  | 29/38 [01:58<00:18,  2.05s/it]2025-11-12 15:12:37 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  79%|███████▉  | 30/38 [01:58<00:12,  1.61s/it]2025-11-12 15:12:37 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  82%|████████▏ | 31/38 [01:58<00:08,  1.23s/it]2025-11-12 15:12:43 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  84%|████████▍ | 32/38 [02:04<00:14,  2.46s/it]2025-11-12 15:12:58 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  87%|████████▋ | 33/38 [02:19<00:31,  6.39s/it]2025-11-12 15:12:59 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  89%|████████▉ | 34/38 [02:20<00:18,  4.71s/it]2025-11-12 15:13:05 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  92%|█████████▏| 35/38 [02:26<00:15,  5.19s/it]2025-11-12 15:13:10 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  95%|█████████▍| 36/38 [02:31<00:10,  5.03s/it]2025-11-12 15:13:10 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating:  97%|█████████▋| 37/38 [02:31<00:03,  3.61s/it]2025-11-12 15:13:29 ll6684-mlaszk.ad.oicr.on.ca httpx[16357] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating: 100%|██████████| 38/38 [02:50<00:00,  4.49s/it]\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'keyword' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 121\u001b[39m\n\u001b[32m    118\u001b[39m df_full = reduce(\u001b[38;5;28;01mlambda\u001b[39;00m left, right: pd.merge(left, right, on=\u001b[33m\"\u001b[39m\u001b[33mpmid\u001b[39m\u001b[33m\"\u001b[39m, how=\u001b[33m\"\u001b[39m\u001b[33mouter\u001b[39m\u001b[33m\"\u001b[39m), data_frames)\n\u001b[32m    120\u001b[39m \u001b[38;5;66;03m# Print Availability Counts\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m121\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mFull Text Availability for keyword \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[43mkeyword\u001b[49m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m:\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    122\u001b[39m availability_counts = df_full[\u001b[33m\"\u001b[39m\u001b[33mFullText_Availability\u001b[39m\u001b[33m\"\u001b[39m].value_counts()\n\u001b[32m    123\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33mFull Text Availability Counts:\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[31mNameError\u001b[39m: name 'keyword' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "# Prints article information for pmids retrieved\n",
    "for pmid in pmids:\n",
    "    article = fetch.article_by_pmid(pmid)\n",
    "    print(f\"PMID {pmid}: {article.title}\")\n",
    "    print(f\"Authors: {', '.join([str(a) for a in article.authors])}\")\n",
    "    print(f\"Journal: {article.journal}\")\n",
    "    print(\"---\")\n",
    "\n",
    "#Generates fulltext-based summaries\n",
    "titles, full_texts = fetch_full_texts(pmids, fetch)\n",
    "filtered_texts = {pmid: text for pmid, text in full_texts.items() if text != \"Full text unavailable\"}\n",
    "# full_text_summaries = await gpt4_summarize_async(filtered_texts, max_length=max_summary_length)\n",
    "# print(full_text_summaries)\n",
    "\n",
    "#Generates abstract-based summaries\n",
    "titles, abstracts = fetch_abstracts(pmids, fetch)\n",
    "# abstract_summaries = await gpt4_summarize_async(abstracts, max_length=max_summary_length)\n",
    "# print(abstract_summaries)\n",
    "\n",
    "links = {pmid: f\"https://pubmed.ncbi.nlm.nih.gov/{pmid}/\" for pmid in pmids}\n",
    "\n",
    "\n",
    "# Calculates similarity between abstract vs full text summaries\n",
    "# Define the evaluate_similarity function\n",
    "def evaluate_similarity(abstract_summaries, full_text_summaries):\n",
    "    vectorizer = TfidfVectorizer()\n",
    "    tfidf_matrix = vectorizer.fit_transform([abstract_summaries, full_text_summaries])\n",
    "    similarity = cosine_similarity(tfidf_matrix[0:1], tfidf_matrix[1:2])\n",
    "    return similarity[0][0]\n",
    "\n",
    "similarity_scores = {\n",
    "    pmid: evaluate_similarity(abstract_summaries[pmid], full_text_summaries[pmid])\n",
    "    for pmid in abstract_summaries\n",
    "    if abstract_summaries[pmid] != \"No text available\"\n",
    "    and full_text_summaries.get(pmid)\n",
    "    and full_text_summaries[pmid] != \"No text available\"\n",
    "}\n",
    "\n",
    "# Calculate readability scores for abstract summaries\n",
    "abstract_readability = {\n",
    "    pmid: compute_readability(abstract_summaries[pmid])\n",
    "    for pmid in abstract_summaries\n",
    "}\n",
    "\n",
    "# Calculate readability scores for full text summaries\n",
    "full_text_readability = {\n",
    "    pmid: compute_readability(full_text_summaries[pmid])\n",
    "    for pmid in full_text_summaries\n",
    "}\n",
    "\n",
    "\n",
    "# Full text availability\n",
    "full_text_availability = {\n",
    "    pmid: \"Available\" if full_texts[pmid] != \"Full text unavailable\" else \"Unavailable\"\n",
    "    for pmid in pmids\n",
    "}\n",
    "\n",
    "# Define the prompt used for summarization\n",
    "ragas_prompt = (\"Summarize this article in 250 to 350 words for a 2nd grade lay audience using simple words.\"\n",
    "                \"The tone should be casual.\"\n",
    "                \"Avoid emotive or negative language such as battle or fight.\"\n",
    "                \"Keep words under 4 syllables and sentences under 10 words.\"\n",
    "                )\n",
    "\n",
    "# Compute RAGAS faithfulness\n",
    "abstract_ragas_faithfulness = compute_ragas_faithfulness(abstract_summaries, abstracts, ragas_prompt)\n",
    "fulltext_ragas_faithfulness = compute_ragas_faithfulness(full_text_summaries, full_texts, ragas_prompt)\n",
    "\n",
    "\n",
    "# Prepare DataFrames\n",
    "Title = pd.DataFrame(list(titles.items()), columns=[\"pmid\", \"Title\"])\n",
    "Link = pd.DataFrame(list(links.items()), columns=[\"pmid\", \"Link\"])\n",
    "Abstract = pd.DataFrame(list(abstracts.items()), columns=[\"pmid\", \"Abstract\"])\n",
    "Abstract = Abstract.assign(\n",
    "    Abstract=lambda df: df[\"Abstract\"].str.replace(\"\\n\", \" \")\n",
    ")\n",
    "FullTextFlag = pd.DataFrame(list(full_text_availability.items()), columns=[\"pmid\", \"FullText_Availability\"])\n",
    "Abstract_Summary = pd.DataFrame(list(abstract_summaries.items()), columns=[\"pmid\", \"Abstract_Summary\"])\n",
    "Full_Text_Summary = pd.DataFrame(list(full_text_summaries.items()), columns=[\"pmid\", \"Full_Text_Summary\"])\n",
    "Similarity = pd.DataFrame(list(similarity_scores.items()), columns=[\"pmid\", \"Abstract_vs_FullText_Similarity\"])\n",
    "\n",
    "Abstract_Readability = pd.DataFrame([\n",
    "    {\"pmid\": pmid, \n",
    "    \"Abstract_Flesch_Reading_Ease\": scores[0], \n",
    "    \"Abstract_Flesch_Kincaid_Grade\": scores[1]}\n",
    "    for pmid, scores in abstract_readability.items()\n",
    "])\n",
    "\n",
    "\n",
    "Full_Text_Readability = pd.DataFrame([\n",
    "    {\"pmid\": pmid, \n",
    "    \"FullText_Flesch_Reading_Ease\": scores[0], \n",
    "    \"FullText_Flesch_Kincaid_Grade\": scores[1]}\n",
    "    for pmid, scores in full_text_readability.items()\n",
    "])\n",
    "\n",
    "# RAGAS faithfulness scores\n",
    "Abstract_RAGAS_Faithfulness = pd.DataFrame(list(abstract_ragas_faithfulness.items()), columns=[\"pmid\", \"Abstract_RAGAS_Faithfulness\"])\n",
    "FullText_RAGAS_Faithfulness = pd.DataFrame(list(fulltext_ragas_faithfulness.items()), columns=[\"pmid\", \"FullText_RAGAS_Faithfulness\"])\n",
    "\n",
    "\n",
    "# Combine all DataFrames\n",
    "data_frames = [\n",
    "    Title, \n",
    "    Link, \n",
    "    Abstract, \n",
    "    FullTextFlag, \n",
    "    Abstract_Summary, \n",
    "    Full_Text_Summary, \n",
    "    Similarity,\n",
    "    Abstract_Readability,\n",
    "    Full_Text_Readability,\n",
    "    Abstract_RAGAS_Faithfulness,\n",
    "    FullText_RAGAS_Faithfulness\n",
    "]\n",
    "\n",
    "# Combine all DataFrames\n",
    "df_full = reduce(lambda left, right: pd.merge(left, right, on=\"pmid\", how=\"outer\"), data_frames)\n",
    "\n",
    "# Print Availability Counts\n",
    "print(f\"Full Text Availability for keyword '{keyword}':\")\n",
    "availability_counts = df_full[\"FullText_Availability\"].value_counts()\n",
    "print(\"Full Text Availability Counts:\")\n",
    "print(availability_counts)\n",
    "\n",
    "# Filter to keep only articles with full text available\n",
    "df_full = df_full[df_full[\"FullText_Availability\"] == \"Available\"]\n",
    "\n",
    "# Save the DataFrame to a CSV file with a timestamp\n",
    "timestamp = datetime.now().strftime(\"%Y%m%d_%H%M%S\")\n",
    "base_name, _ = os.path.splitext(output_file)\n",
    "output_file_with_timestamp = f\"{base_name}_{timestamp}.csv\"\n",
    "\n",
    "# Specify the output directory\n",
    "output_dir = Path(\"./output\")\n",
    "output_dir.mkdir(parents=True, exist_ok=True)\n",
    "full_output_path = output_dir / output_file_with_timestamp\n",
    "df_full.to_csv(full_output_path, index=False)\n",
    "\n",
    "print(f\"Data for keyword '{keyword}' saved to: {output_file_with_timestamp}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5734cecf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/4s/6g39d69s7m35v6gzfsln6thr0000gp/T/ipykernel_7485/1854503017.py:105: DeprecationWarning: LangchainLLMWrapper is deprecated and will be removed in a future version. Use the modern LLM providers instead: from ragas.llms.base import llm_factory; llm = llm_factory('gpt-4o-mini') or from ragas.llms.base import instructor_llm_factory; llm = instructor_llm_factory('openai', client=openai_client)\n",
      "  evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model=\"gpt-4o-mini\"))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LangchainLLMWrapper(langchain_llm=ChatOpenAI(...))\n",
      "RAGAS data: {'question': ['Summarize this article in 250 to 350 words for a 2nd grade lay audience using simple words.The tone should be casual.Avoid emotive or negative language such as battle or fight.Keep words under 4 syllables and sentences under 10 words.'], 'answer': ['This work looks at drugs for kidney cancer.  \\nOne drug, everolimus, helps at first.  \\nBut it can stop working over time.  \\nThe team looked for new ways to help.  \\nThey tested a second drug, SGI-1027.  \\nSGI-1027 can change how DNA acts.  \\nIt makes cells fill with big bubbles.  \\nThe bubbles can make the cell break.  \\nThe team found SGI-1027 helps everolimus.  \\nThe two drugs work well when used together.  \\nThey slow tumor cell growth and move.  \\nThey cut how cells spread and invade.  \\nThe mix makes cells die in two ways.  \\nOne way is a calm, planned shut down.  \\nThe other way makes cells swell and pop.  \\nA protein called GSDME joins the pop way.  \\nKidney cancer cells had more GSDME inside.  \\nTheir lysosome bags were also more busy.  \\nLysosomes are the cell clean up bags.  \\nThese bags can get leaky from the drugs.  \\nThe leaks can push cells to die fast.  \\nThis gives a clear chance to treat cells.  \\nThe drug mix worked well in small mice tests.  \\nTumors were small and the mice did fine.  \\nNo big harm showed with the mix dose.  \\nThis study shows SGI-1027 can kill tumor cells.  \\nIt also shows SGI-1027 can help everolimus work.  \\nThe find may help when everolimus stops working.  \\nMore tests are needed to see it in people.  \\nBut this work gives new hope for kidney care.'], 'contexts': [['The mTOR inhibitor everolimus has been approved as a sequential or second-line therapy for renal cell carcinoma (RCC). However, the development of drug resistance limits its clinical applications. This study aims to address the challenge of everolimus resistance and provide new insights into the treatment of advanced RCC. Here, the cytotoxicity of the DNA methyltransferase 1 (DNMT1) inhibitor SGI-1027 in inducing cell vacuolation and methuosis is discovered and demonstrated for the first time. Additionally, SGI-1027 exerts synergistic effects with everolimus, as their combination suppresses the growth, migration, and invasion of renal cancer cells. Mechanistically, apoptosis and GSDME-dependent pyroptosis triggered by lysosomal membrane permeability (LMP) are observed. The upregulation of GSDME expression and increased lysosomal activity in renal cancer cells provide a therapeutic window for the combination of these two drugs to treat renal cancer. The combination treatment exhibits effective anti-tumor activity and is well tolerated in a subcutaneous tumor model. Overall, this study validates and reveals the specific cytotoxicity property of SGI-1027 and its potent synergistic effect with everolimus, offering new insights into advanced RCC therapy and everolimus-resistance overcoming.']]}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Evaluating:   0%|          | 0/1 [00:00<?, ?it/s]2025-11-11 00:29:31 ML8136-T2YU.local httpx[7485] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "2025-11-11 00:30:06 ML8136-T2YU.local httpx[7485] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "Evaluating: 100%|██████████| 1/1 [00:44<00:00, 44.20s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RAGAS results: {'faithfulness': 0.9333}\n",
      "<openai.OpenAI object at 0x34def3890>\n"
     ]
    }
   ],
   "source": [
    "test = compute_ragas_faithfulness(abstract_summaries, abstracts, ragas_prompt)\n",
    "print(client)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
